<!DOCTYPE html>
<html lang="en-GB">

<head>
	<meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="pingback" href="https://samnicholls.net/xmlrpc.php">

	
	<title>What am I doing? &#8211; Samposium</title>
<meta name='robots' content='max-image-preview:large' />
<link rel='dns-prefetch' href='//s0.wp.com' />
<link rel='dns-prefetch' href='//fonts.googleapis.com' />
<link rel='dns-prefetch' href='//s.w.org' />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Feed" href="https://samnicholls.net/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Comments Feed" href="https://samnicholls.net/comments/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; What am I doing? Comments Feed" href="https://samnicholls.net/2015/04/27/what-am-i-doing/feed/" />
		<script type="text/javascript">
			window._wpemojiSettings = {"baseUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/72x72\/","ext":".png","svgUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/svg\/","svgExt":".svg","source":{"concatemoji":"https:\/\/samnicholls.net\/wp-includes\/js\/wp-emoji-release.min.js?ver=5.7.5"}};
			!function(e,a,t){var n,r,o,i=a.createElement("canvas"),p=i.getContext&&i.getContext("2d");function s(e,t){var a=String.fromCharCode;p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,e),0,0);e=i.toDataURL();return p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,t),0,0),e===i.toDataURL()}function c(e){var t=a.createElement("script");t.src=e,t.defer=t.type="text/javascript",a.getElementsByTagName("head")[0].appendChild(t)}for(o=Array("flag","emoji"),t.supports={everything:!0,everythingExceptFlag:!0},r=0;r<o.length;r++)t.supports[o[r]]=function(e){if(!p||!p.fillText)return!1;switch(p.textBaseline="top",p.font="600 32px Arial",e){case"flag":return s([127987,65039,8205,9895,65039],[127987,65039,8203,9895,65039])?!1:!s([55356,56826,55356,56819],[55356,56826,8203,55356,56819])&&!s([55356,57332,56128,56423,56128,56418,56128,56421,56128,56430,56128,56423,56128,56447],[55356,57332,8203,56128,56423,8203,56128,56418,8203,56128,56421,8203,56128,56430,8203,56128,56423,8203,56128,56447]);case"emoji":return!s([55357,56424,8205,55356,57212],[55357,56424,8203,55356,57212])}return!1}(o[r]),t.supports.everything=t.supports.everything&&t.supports[o[r]],"flag"!==o[r]&&(t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&t.supports[o[r]]);t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&!t.supports.flag,t.DOMReady=!1,t.readyCallback=function(){t.DOMReady=!0},t.supports.everything||(n=function(){t.readyCallback()},a.addEventListener?(a.addEventListener("DOMContentLoaded",n,!1),e.addEventListener("load",n,!1)):(e.attachEvent("onload",n),a.attachEvent("onreadystatechange",function(){"complete"===a.readyState&&t.readyCallback()})),(n=t.source||{}).concatemoji?c(n.concatemoji):n.wpemoji&&n.twemoji&&(c(n.twemoji),c(n.wpemoji)))}(window,document,window._wpemojiSettings);
		</script>
		<style type="text/css">
img.wp-smiley,
img.emoji {
	display: inline !important;
	border: none !important;
	box-shadow: none !important;
	height: 1em !important;
	width: 1em !important;
	margin: 0 .07em !important;
	vertical-align: -0.1em !important;
	background: none !important;
	padding: 0 !important;
}
</style>
	<link rel='stylesheet' id='crayon-css'  href='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/css/min/crayon.min.css?ver=_2.7.2_beta' type='text/css' media='all' />
<link rel='stylesheet' id='wp-block-library-css'  href='https://samnicholls.net/wp-includes/css/dist/block-library/style.min.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='child-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='lato-css'  href='//fonts.googleapis.com/css?family=Lato%3A400%2C700%2C400italic&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='slabo-css'  href='//fonts.googleapis.com/css?family=Slabo+27px&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='fa-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/font-awesome.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='rd-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-custom-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/custom.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='jetpack_css-css'  href='https://samnicholls.net/wp-content/plugins/jetpack/css/jetpack.css?ver=4.9' type='text/css' media='all' />
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery.min.js?ver=3.5.1' id='jquery-core-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery-migrate.min.js?ver=3.3.2' id='jquery-migrate-js'></script>
<script type='text/javascript' id='crayon_js-js-extra'>
/* <![CDATA[ */
var CrayonSyntaxSettings = {"version":"_2.7.2_beta","is_admin":"0","ajaxurl":"https:\/\/samnicholls.net\/wp-admin\/admin-ajax.php","prefix":"crayon-","setting":"crayon-setting","selected":"crayon-setting-selected","changed":"crayon-setting-changed","special":"crayon-setting-special","orig_value":"data-orig-value","debug":""};
var CrayonSyntaxStrings = {"copy":"Press %s to Copy, %s to Paste","minimize":"Click To Expand Code"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/js/min/crayon.min.js?ver=_2.7.2_beta' id='crayon_js-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/fd-footnotes/fdfootnotes.js?ver=1.34' id='fdfootnote_script-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/bootstrap.js?ver=5.7.5' id='bootstrap-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/custom.js?ver=5.7.5' id='js-custom-js'></script>
<link rel="https://api.w.org/" href="https://samnicholls.net/wp-json/" /><link rel="alternate" type="application/json" href="https://samnicholls.net/wp-json/wp/v2/posts/269" /><link rel="EditURI" type="application/rsd+xml" title="RSD" href="https://samnicholls.net/xmlrpc.php?rsd" />
<link rel="wlwmanifest" type="application/wlwmanifest+xml" href="https://samnicholls.net/wp-includes/wlwmanifest.xml" /> 
<meta name="generator" content="WordPress 5.7.5" />
<link rel="canonical" href="https://samnicholls.net/2015/04/27/what-am-i-doing/" />
<link rel='shortlink' href='https://wp.me/p6RfP0-4l' />
<link rel="alternate" type="application/json+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2015%2F04%2F27%2Fwhat-am-i-doing%2F" />
<link rel="alternate" type="text/xml+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2015%2F04%2F27%2Fwhat-am-i-doing%2F&#038;format=xml" />
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-62206786-1', 'auto');
  ga('send', 'pageview');

</script>


<link rel='dns-prefetch' href='//v0.wordpress.com'>
<style type='text/css'>img#wpstats{display:none}</style>
<!-- Jetpack Open Graph Tags -->
<meta property="og:type" content="article" />
<meta property="og:title" content="What am I doing?" />
<meta property="og:url" content="https://samnicholls.net/2015/04/27/what-am-i-doing/" />
<meta property="og:description" content="A week ago I had a progress meeting with Amanda and Wayne, who make up the supervisory team for the computational face of my project. I talked about how computers are terrible and where the projectâ€¦" />
<meta property="article:published_time" content="2015-04-27T10:14:14+00:00" />
<meta property="article:modified_time" content="2016-01-13T00:06:37+00:00" />
<meta property="og:site_name" content="Samposium" />
<meta property="og:image" content="https://s0.wp.com/i/blank.jpg" />
<meta property="og:locale" content="en_GB" />
<meta name="twitter:card" content="summary" />
</head>

<body class="post-template-default single single-post postid-269 single-format-standard">


        
<header id="main-header">
    <nav class="navbar-default navbar-static-top navbar" role="navigation">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#rd-collapse">
                    <span class="fa fa-bars"></span>
                    <span class="sr-only screen-reader-text">Toggle navigation</span>
                </button>
                <a class="navbar-brand" href=" https://samnicholls.net/">Samposium</a>            </div>
            
            <div id="rd-collapse" class="collapse navbar-collapse">
            
            <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body class="post-template-default single single-post postid-269 single-format-standard">

<div id="main-menu" class="navbar-nav nav"><ul>
<li class="page_item page-item-8"><a href="https://samnicholls.net/about/">About Me</a></li>
<li class="page_item page-item-11"><a href="https://samnicholls.net/talks/">Talks</a></li>
<li class="page_item page-item-15"><a href="https://samnicholls.net/things-i-have-learned/">Things I Have Learned</a></li>
</ul></div></body></html>
        </div>

        </div>
    </nav>
</header>

<div class="container">

    <div class="row"><!-- Content -->
<section class="col-md-8 content-container">
                          <article class="card post hentry post-269 type-post status-publish format-standard category-au-phd tag-introduction tag-phd tag-project" id="post-269">
                <header class="entry-header">
                        <h1 class="entry-title">What am I doing?</h1>                </header><!-- .entry-header -->

                <div class="meta-well">
                    <span style="margin-right:15px;" class=""><span class="fa fa-user"></span>&nbsp;&nbsp;&nbsp;<span class="author vcard"><a class="url fn n" href="https://samnicholls.net/about/">Sam Nicholls</a></span></span>

                    <span style="margin-right:15px;" class="entry-date published updated post-date"><span class="fa fa-clock-o"></span>&nbsp;&nbsp; <time class="entry-date published" datetime="2015-04-27T11:14:14+01:00">27th April 2015</time><time class="updated" datetime="2016-01-13T00:06:37+00:00">13th January 2016</time></span>                    <span style="margin-right:15px;" class=""><span class="fa fa-comments-o"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/2015/04/27/what-am-i-doing/#respond"><span class="dsq-postid" data-dsqidentifier="269 http://blog.ironowl.io/?p=269">No Comments yet</span></a></span>
                    <span class=""><span class="fa fa-folder-open"></span>&nbsp;&nbsp;&nbsp;<a href="https://samnicholls.net/category/bioinf/au-phd/" rel="category tag">AU-PhD</a></span>
                </div>

                <div class="entry-content"><p>A week ago I had a progress meeting with <a href="http://www.aber.ac.uk/en/cs/staff-list/staff_profiles/?staff_id=afc">Amanda</a> and <a href="http://www.aber.ac.uk/en/cs/staff-list/staff_profiles/?login=waa2">Wayne</a>, who make up the supervisory team for the computational face of my project. I talked about how computers are terrible and where the project is heading.</p>
<p>As Wayne had been away from meetings for a few weeks, I began with a roundup of everything that has been going <a href="/2015/04/17/exits/">disasterously wrong</a><sup id="fnref-269-1"><a href="#fn-269-1" rel="footnote">1</a></sup>. Progress on a functional analysis of the <a href="/2015/04/21/the-story-so-far-p1/">limpet data</a> has been repeatedly hindered by a lack of resources on our cluster which is simply strugging with the sheer size of the jobs I&#8217;m asking of it.</p>
<h1>The Cluster Conundrum</h1>
<p>I&#8217;ve encountered two main issues with job size here:</p>
<ul>
<li>Jobs that are large because the inputs are large but few (<em>e.g.</em> assembling raw reads contained in a pair of 42GB files), or</li>
<li>Jobs that are large because although the inputs are small (&lt; 100MB), there are thousands of them (<em>e.g.</em> <code>BLAST</code>&#8216;ing large numbers of contigs against a sharded database<sup id="fnref-269-2"><a href="#fn-269-2" rel="footnote">2</a></sup>)</li>
</ul>
<h2>Small-Big Jobs</h2>
<p>The former is somewhat unavoidable. If <code>velvet</code> wants to consume 450GB of RAM for an assembly and we want an assembly specifically from <code>velvet</code> then it&#8217;s a case of having to wait patiently for one of the larger nodes to become free enough to schedule the job. Although we could look for other assemblers<sup id="fnref-269-3"><a href="#fn-269-3" rel="footnote">3</a></sup> and evaluate their bold claims regarding reduced resource usage over competitors often when we&#8217;ve found a tool that <em>just works</em>, we like to keep things that way &#8212; especially if we want to be able to compare results of other assemblies that must be manufactured in the same way.</p>
<p>Cluster jobs require resources to be requested up front and guesstimating (even generously) can often lead to a job being terminated for exceeding its allowance, wasting queue time (days) as well as execution time (days or weeks) and leaving you with nothing to show<sup id="fnref-269-4"><a href="#fn-269-4" rel="footnote">4</a></sup>. The problem is in asking for too much, you queue for a node longer, but when finally scheduled you effectively block others from using resources for a significant time period and <a href="/2015/04/26/memblame/">I&#8217;ll make you feel bad for it</a>.</p>
<p>The only way to get around these constraints is to minimise the dataset you have in the first place. For example for assemblies you could employ:</p>
<p><strong>Normalization</strong> : Count appearances of substrings of length <em>k</em> (<strong>k-mers</strong>) present in the raw reads, then discard corresponding reads in a fashion that retains the distribution of k-mers. Discarding data is clearly lossy, but the idea is that as the distribution of k-mers is represented in the same way but with fewer reads.</p>
<p><strong>Partitioning</strong> : Attempt to construct a graph of all k-mers present in the raw reads, then partition it in to a series of subgraphs based on connectivity. Corresponding reads from each partition can then be assembled separately and potentially merged afterwards. Personally I&#8217;ve found this method a bit hit and miss so far but would like to have time to investigate further.</p>
<p><strong>Subsampling</strong> : Select a more manageable proportion of reads from your dataset at random and construct an assembly. Not only very lossy, this in itself raises some interesting sampling bias issues (to go with your original environment sampling and PCR biases).</p>
<p><strong>Iterative Subsampling</strong> : Assemble a subsample from your data set and then align the contigs back to the original raw reads. Re-subsample from all remaining unaligned reads and create a second assembly, repeat the process until you have <em>N</em> different assemblies and are satisfied with the overall alignment (<em>i.e.</em> the set of remaining unaligned reads is sufficiently small). Tom in our lab group has been pioneering this approach and might hopefully give a better explanation of this than I can.</p>
<h2>Big-Small Jobs</h2>
<p>The latter category is a problem actually introduced by trying to optimise cluster scheduling in the first place. For example, an assembly can produce thousands of <strong>contigs</strong> groups of reads believed by an assembler to belong together and often we want to know if any interesting known sequences can be found on these contigs. Databases of interesting known sequences are often <a href="/2015/04/24/trembling/">(very) large</a> and so to avoid submitting an inefficient long-running memory-hogging small-big job to locate thousands of different needles in thousands of different haystacks (<em>i.e.</em> <code>BLAST</code>&#8216;ing many contigs against a large database), we can instead attempt to minimize the size of the job by amortising the work over many significantly smaller jobs.</p>
<p>For the purpose of <code>BLAST</code><sup id="fnref-269-5"><a href="#fn-269-5" rel="footnote">5</a></sup>, we can shard both the contigs and the database of interesting sequences in to smaller pieces. This reduces the search space (fewer interesting-sequence needles to find in fewer contig haystacks) and thus execution time and resource requirements. Now your monolith job is represented by hundreds (or thousands) of smaller, less resource intensive jobs that finish more quickly. Hooray!</p>
<p>Until the number of jobs you have starts <a href="/2015/04/17/exits/">causing trouble</a>.</p>
<p>Of course this in turn makes handling data for downstream analysis a little more complex, output files need <a href="/2015/04/29/pipelines/">converting, sorting and merging</a> before potentially having to be re-sharded once again to fit them through a different tool.</p>
<h1>Conquering Complications</h1>
<p>So how can we move forward? We could just do what is <a href="https://xkcd.com/927/">fashionable at the moment</a> and write a fantastic new [assembler|aligner|pipeline] that is better and faster<sup id="fnref-269-6"><a href="#fn-269-6" rel="footnote">6</a></sup> than the competition, uses next-to-no memory and can even be run on a Raspberry Pi, but this is more than a PhD in itself<sup id="fnref-269-7"><a href="#fn-269-7" rel="footnote">7</a></sup>, so sadly, I guess we have to make do and stick with what we have and attempt to use it more efficiently.</p>
<p>Digressing, I feel a major problem in bioinformatics software right now is a failure to adequately communicate the uses and effects of parameters: how can end-users of your software fine tune<sup id="fnref-269-8"><a href="#fn-269-8" rel="footnote">8</a></sup> controls and options without it feeling like piloting a Soyuz? I think if the learning curve is too great, with understanding hampered further by a lack of tutorials or extensive documentation with examples, users end up driven to roll their own solution. Often in these cases the end result is maintained by a single developer or group, missing out on the benefits of input from the open-source community at large.</p>
<p><em>Small-Big</em> jobs can currently be tackled with novel methods like Tom&#8217;s iterative subsampling as described above, or of course, by adding additional resources (but that costs money).</p>
<p>Some of the risk recently identified with the execution of <em>Big-Small</em> jobs can be reduced by being a little more organised. I&#8217;m in the process of writing some software to ease interaction with Sun Grid Engine that now places logs generated during job execution outside of the working directory &#8212; reducing some of the I/O load when repeatedly requesting the contents of output directories.</p>
<p>Keeping abreast of the work of others who dared to tread and write their own new assembler, aligner or whatever is important too. Currently we&#8217;re testing out <a href="http://omics.informatics.indiana.edu/mg/RAPSearch2/"><code>rapsearch</code></a> as an alternative to <code>BLAST</code> simply due to its execution speed (yet another post in itself). <code>BLAST</code> is pretty old and &#8220;better&#8221; alternatives are known to exist, but it&#8217;s still oft-cited and an expected part of analysis in journal papers, so switching out parts of our pipeline for performance is not ideal. At the same time, I actually want to get some work done and right now using <code>BLAST</code> on the dataset I have, with the resources I have is proving too problematic.</p>
<p>At the very least, we can now use <code>rapsearch</code> to quickly look for hits to be analysed further with <code>BLAST</code> if we fear that the community may be put off by our use of &#8220;non-standard&#8221; software.</p>
<h1>Ignoring the Impossible</h1>
<p>After trading some graph theory with Wayne in return for some biological terminology, we turned our attention to a broad view of where the project as whole is heading. We discussed how it is difficult to assemble entire genomes from metagenomic datasets due to environmental bias, PCR bias and clearly, computational troubles.</p>
<p>I&#8217;d described my project at a talk previously:</p>
<blockquote><p>
[&#8230;] it&#8217;s like trying to simultaneously assemble thousands of jigsaws but some of the jigsaws are heavily duplicated and some of the jigsaws hardly appear at all, a lot of the pieces are missing and quite a few pieces that really should fit together are broken. Also the jigsaws are pictures of sky.
</p></blockquote>
<p>Lately I&#8217;ve started to wonder how this is even possible: how can we state with confidence that we&#8217;ve assembled a whole environment? How do we know the initial sample contained all the species? How can we determine what is sequencing error and what is real and rare? How on Earth are we supposed to identify all affinities in variation for all species across millions of reads that are shorter than my average <a href="https://twitter.com/samstudio8/">Tweet</a> that barely overlap?</p>
<p>We can&#8217;t<sup id="fnref-269-9"><a href="#fn-269-9" rel="footnote">9</a></sup>.</p>
<p>But that&#8217;s ok. That isn&#8217;t the project. These sorts of aims are too broad, though that won&#8217;t prevent me from trying. Currently I&#8217;m hunting for <strong>hydrolases</strong> (enzymes used to break apart chemical bonds in presence of water), so we can turn the problem on its head a little. Instead of creating an assembly and assigning taxonomic and functional annotations to every single one of the resulting contigs then filtering the results by those that resemble hydrolasic behaviour &#8211; treating each contig as equally interesting &#8211; we can just look at contigs that contain coding regions for the creation of hydrolases directly! We can use a short-read aligner such as <code>rapsearch</code> or <code>BLAST</code> to search for needles from a hydrolase-specific database of our own construction, instead of a larger, more general bacterial database.</p>
<p>We can then query the assembly for the original raw reads that built the contig on which strong hits for hydrolases appear. We can take a closer look at these reads alone, filtering out whole swathes of the assembly (and thus millions of reads) that are &#8220;uninteresting&#8221; in terms of our search.</p>
<p>We want to identify and extract interesting enzymes and the sequences that derive them, discovering a novel species in the process is a nice bonus but the protein sequence is the key.</p>
<hr />
<h1>tl;dr</h1>
<ul>
<li>My data is too big and my computer is too small.</li>
<li>There are <em>big-small</em> jobs and <em>small-big</em> jobs and both are problematic and unavoidable.</li>
<li>There just isn&#8217;t time to look at everything that is interesting.</li>
<li>We need to know the tools we are using inside out and have a very good reason to make our own.</li>
<li>We don&#8217;t have to care about data that we aren&#8217;t interested in.</li>
<li>The project <strong>probably</strong> isn&#8217;t impossible.</li>
</ul>
<div class="footnotes">
<hr />
<ol>
<li id="fn-269-1">
Which is pretty much anything that involves a computer.&#160;<a href="#fnref-269-1" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-2">
In an attempt to speed up BLAST queries against large databases we have taken to splitting the database into &#8216;shards&#8217;; submitting a job for each set of contigs against a specific database shard, before <code>cat</code>&#8216;ing all the results together at the end. I call this <strong>re-tailing</strong>.&#160;<a href="#fnref-269-2" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-3">
In fact, currently I&#8217;m trying to evaluate <a href="https://github.com/voutcn/megahit"><code>MegaHIT</code></a>.&#160;<a href="#fnref-269-3" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-4">
This isn&#8217;t always strictly true. For example, aligners can flush output hits to a file as they go along and with a bit of fiddling you can pick up where you left off and <code>cat</code> the outputs together<sup id="fnref2:269-5"><a href="#fn-269-5" rel="footnote">5</a></sup>.&#160;<a href="#fnref-269-4" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-5">
Other short-read sequencer aligners are available.&#160;<a href="#fnref-269-5" rev="footnote">&#8617;</a> <a href="#fnref2:269-5" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-6">
Bonus points for ensuring it is also harder and stronger.&#160;<a href="#fnref-269-6" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-7">
I learned from my undergraduate dissertation that no matter how hard you try, the time to investigate every interesting side-street simply does not exist and it&#8217;s important to try and stay on some form of track.&#160;<a href="#fnref-269-7" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-8">
I had a brief discussion about the difficulty of automated parameter selection on Twitter after a virtual conference and this is something I&#8217;d like to write more about at length in future:</p>
<blockquote class="twitter-tweet" lang="en"><p>.<a href="https://twitter.com/bioinformatics">@bioinformatics</a> nucleotid.es for biologists &amp; not to document &quot;fine tuning&quot;. But most assemblers need this tuning? <a href="https://twitter.com/hashtag/BaltiAndBioinformatics?src=hash">#BaltiAndBioinformatics</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/557972285014155267">January 21, 2015</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p><a href="#fnref-269-8" rev="footnote">&#8617;</a>
</li>
<li id="fn-269-9">
Probably.&#160;<a href="#fnref-269-9" rev="footnote">&#8617;</a>
</li>
</ol>
</div>
</div>
                
                        <div class="tag-well"><span><span class="fa fa-tags"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/tag/introduction/" rel="tag">introduction</a>,&nbsp; <a href="https://samnicholls.net/tag/phd/" rel="tag">phd</a>,&nbsp; <a href="https://samnicholls.net/tag/project/" rel="tag">project</a></span></div>
                
                <nav class="further-reading">
	<div class="previous">
		<span>Previous Post</span>
		<a href="https://samnicholls.net/2015/04/26/memblame/">`memblame`</a>
	</div>
	<div class="next">
		<span>Next Post</span>
		<a href="https://samnicholls.net/2015/04/29/pipelines/">Pipelines</a>
	</div>
</nav>            </article>
                                <div class="card">

<div id="disqus_thread"></div>
                </div>
    
</section>

        <!-- Sidebar -->
        <aside class="col-md-4 sidebar-container">
		            <div id="widget-area" class="widget-area" role="complementary">
                <div id="custom_html-3" class="widget_text card widget widget_custom_html"><div class="textwidget custom-html-widget"><img src="https://pbs.twimg.com/profile_images/1044255545610436609/wt33ZbtU_400x400.jpg" height=120 style="margin-top: 20px; margin-right:10px; float: left;" alt="Sam Nicholls" />
<div style="padding-top:15px; line-height: normal"><b style="font-size: 1.2em">Sam Nicholls</b><br><span style="font-size: 0.9em;">Haplotype Wrangler at University of Birmingham.<br>I tell computers to do things, then I write about how it all went wrong.</span></div></div></div><div id="search-2" class="card widget widget_search"><form style="margin-top:15px" action="https://samnicholls.net/" id="searchform" method="get" role="form">
    <div class="form-group">
        <input type="text" class="form-control" name="s" id="s" placeholder="e.g. Search...">
    </div>
</form>
</div><div id="twitter_timeline-5" class="card widget widget_twitter_timeline"><a class="twitter-timeline" data-height="400" data-theme="light" data-link-color="#f96e5b" data-border-color="#e8e8e8" data-lang="EN" data-partner="jetpack" data-chrome="noheader nofooter noborders" href="https://twitter.com/samstudio8">My Tweets</a></div><div id="categories-5" class="card widget widget_categories"><h2>Categories</h2>
			<ul>
					<li class="cat-item cat-item-2"><a href="https://samnicholls.net/category/bioinf/">Bioinformatics</a>
<ul class='children'>
	<li class="cat-item cat-item-8"><a href="https://samnicholls.net/category/bioinf/au-phd/">AU-PhD</a>
</li>
	<li class="cat-item cat-item-3"><a href="https://samnicholls.net/category/bioinf/sanger-qc/" title="Sanger QC Project">Sanger-QC</a>
</li>
	<li class="cat-item cat-item-61"><a href="https://samnicholls.net/category/bioinf/tools/">Tools</a>
	<ul class='children'>
	<li class="cat-item cat-item-185"><a href="https://samnicholls.net/category/bioinf/tools/chitin/">chitin</a>
</li>
	</ul>
</li>
</ul>
</li>
	<li class="cat-item cat-item-80"><a href="https://samnicholls.net/category/devops/">Devops</a>
</li>
	<li class="cat-item cat-item-114"><a href="https://samnicholls.net/category/event/">Event</a>
</li>
	<li class="cat-item cat-item-21"><a href="https://samnicholls.net/category/gadgets/">Gadgets</a>
</li>
	<li class="cat-item cat-item-176"><a href="https://samnicholls.net/category/hardware/">Hardware</a>
<ul class='children'>
	<li class="cat-item cat-item-223"><a href="https://samnicholls.net/category/hardware/beehive-monitor/">Beehive Monitor</a>
</li>
	<li class="cat-item cat-item-178"><a href="https://samnicholls.net/category/hardware/esp8266/">ESP8266</a>
</li>
	<li class="cat-item cat-item-236"><a href="https://samnicholls.net/category/hardware/lego-sequencer/">Lego Sequencer</a>
</li>
	<li class="cat-item cat-item-177"><a href="https://samnicholls.net/category/hardware/lightstick/">Lightstick</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-151"><a href="https://samnicholls.net/category/lab/">Lab</a>
<ul class='children'>
	<li class="cat-item cat-item-150"><a href="https://samnicholls.net/category/lab/protocol-guides/">Protocol Guides</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-5"><a href="https://samnicholls.net/category/meta/">Meta</a>
</li>
	<li class="cat-item cat-item-62"><a href="https://samnicholls.net/category/mysteries/">Mysteries</a>
</li>
	<li class="cat-item cat-item-22"><a href="https://samnicholls.net/category/personal/">Personal</a>
</li>
	<li class="cat-item cat-item-4"><a href="https://samnicholls.net/category/photo/">Photography</a>
</li>
	<li class="cat-item cat-item-143"><a href="https://samnicholls.net/category/programming/">Programming</a>
<ul class='children'>
	<li class="cat-item cat-item-145"><a href="https://samnicholls.net/category/programming/documentation/">Documentation</a>
</li>
	<li class="cat-item cat-item-144"><a href="https://samnicholls.net/category/programming/python/">Python</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-48"><a href="https://samnicholls.net/category/status-report/">Status Report</a>
</li>
	<li class="cat-item cat-item-60"><a href="https://samnicholls.net/category/sysadm/">System Administration</a>
</li>
	<li class="cat-item cat-item-1"><a href="https://samnicholls.net/category/uncategorised/">Uncategorised</a>
</li>
			</ul>

			</div>            </div>
		        </aside>
    </div>

</div> <!-- /container -->

<footer id="main-footer" class="bg-primary">
	<div id="footer-widgets" class="container">
		    </div>
</footer>



	<div style="display:none">
	</div>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/comment-reply.min.js?ver=5.7.5' id='comment-reply-js'></script>
<script type='text/javascript' id='disqus_count-js-extra'>
/* <![CDATA[ */
var countVars = {"disqusShortname":"samposium-blog"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_count.js?ver=3.0.17' id='disqus_count-js'></script>
<script type='text/javascript' id='disqus_embed-js-extra'>
/* <![CDATA[ */
var embedVars = {"disqusConfig":{"integration":"wordpress 3.0.17"},"disqusIdentifier":"269 http:\/\/blog.ironowl.io\/?p=269","disqusShortname":"samposium-blog","disqusTitle":"What am I doing?","disqusUrl":"https:\/\/samnicholls.net\/2015\/04\/27\/what-am-i-doing\/","postId":"269"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_embed.js?ver=3.0.17' id='disqus_embed-js'></script>
<script type='text/javascript' src='https://s0.wp.com/wp-content/js/devicepx-jetpack.js?ver=202335' id='devicepx-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/jetpack/_inc/twitter-timeline.js?ver=4.0.0' id='jetpack-twitter-timeline-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/wp-embed.min.js?ver=5.7.5' id='wp-embed-js'></script>
<script type='text/javascript' src='https://stats.wp.com/e-202335.js' async defer></script>
<script type='text/javascript'>
	_stq = window._stq || [];
	_stq.push([ 'view', {v:'ext',j:'1:4.9',blog:'101350222',post:'269',tz:'1',srv:'samnicholls.net'} ]);
	_stq.push([ 'clickTrackerInit', '101350222', '269' ]);
</script>

<script type="text/javascript">
(function() {
  if (navigator.userAgent.match(/IEMobile\/10\.0/)) {
    var msViewportStyle = document.createElement("style");
    msViewportStyle.appendChild(
      document.createTextNode("@-ms-viewport{width:auto!important}")
    );
    document.getElementsByTagName("head")[0].appendChild(msViewportStyle);
  }
})();

var collapsed = false;
jQuery( document ).ready(function($) {
    $('.navbar-toggle').click(function(){
        console.log(collapsed);
        if(collapsed) {
            $('.navbar-header > button').html('<i class="fa fa-bars"></i>');
        } else {
            $('.navbar-header > button').html('<i class="fa fa-times"></i>');
        }
        $('.navbar-collapse').toggle(200);
        collapsed = !collapsed;
    });
});
</script>
	
</body>

</html>
