<!DOCTYPE html>
<html lang="en-GB">

<head>
	<meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="pingback" href="https://samnicholls.net/xmlrpc.php">

	
	<title>Bioinformatics is a disorganised disaster and I am too. So I made a shell. &#8211; Samposium</title>
<meta name='robots' content='max-image-preview:large' />
<link rel='dns-prefetch' href='//s0.wp.com' />
<link rel='dns-prefetch' href='//fonts.googleapis.com' />
<link rel='dns-prefetch' href='//s.w.org' />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Feed" href="https://samnicholls.net/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Comments Feed" href="https://samnicholls.net/comments/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Bioinformatics is a disorganised disaster and I am too. So I made a shell. Comments Feed" href="https://samnicholls.net/2016/11/16/disorganised-disaster/feed/" />
		<script type="text/javascript">
			window._wpemojiSettings = {"baseUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/72x72\/","ext":".png","svgUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/svg\/","svgExt":".svg","source":{"concatemoji":"https:\/\/samnicholls.net\/wp-includes\/js\/wp-emoji-release.min.js?ver=5.7.5"}};
			!function(e,a,t){var n,r,o,i=a.createElement("canvas"),p=i.getContext&&i.getContext("2d");function s(e,t){var a=String.fromCharCode;p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,e),0,0);e=i.toDataURL();return p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,t),0,0),e===i.toDataURL()}function c(e){var t=a.createElement("script");t.src=e,t.defer=t.type="text/javascript",a.getElementsByTagName("head")[0].appendChild(t)}for(o=Array("flag","emoji"),t.supports={everything:!0,everythingExceptFlag:!0},r=0;r<o.length;r++)t.supports[o[r]]=function(e){if(!p||!p.fillText)return!1;switch(p.textBaseline="top",p.font="600 32px Arial",e){case"flag":return s([127987,65039,8205,9895,65039],[127987,65039,8203,9895,65039])?!1:!s([55356,56826,55356,56819],[55356,56826,8203,55356,56819])&&!s([55356,57332,56128,56423,56128,56418,56128,56421,56128,56430,56128,56423,56128,56447],[55356,57332,8203,56128,56423,8203,56128,56418,8203,56128,56421,8203,56128,56430,8203,56128,56423,8203,56128,56447]);case"emoji":return!s([55357,56424,8205,55356,57212],[55357,56424,8203,55356,57212])}return!1}(o[r]),t.supports.everything=t.supports.everything&&t.supports[o[r]],"flag"!==o[r]&&(t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&t.supports[o[r]]);t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&!t.supports.flag,t.DOMReady=!1,t.readyCallback=function(){t.DOMReady=!0},t.supports.everything||(n=function(){t.readyCallback()},a.addEventListener?(a.addEventListener("DOMContentLoaded",n,!1),e.addEventListener("load",n,!1)):(e.attachEvent("onload",n),a.attachEvent("onreadystatechange",function(){"complete"===a.readyState&&t.readyCallback()})),(n=t.source||{}).concatemoji?c(n.concatemoji):n.wpemoji&&n.twemoji&&(c(n.twemoji),c(n.wpemoji)))}(window,document,window._wpemojiSettings);
		</script>
		<style type="text/css">
img.wp-smiley,
img.emoji {
	display: inline !important;
	border: none !important;
	box-shadow: none !important;
	height: 1em !important;
	width: 1em !important;
	margin: 0 .07em !important;
	vertical-align: -0.1em !important;
	background: none !important;
	padding: 0 !important;
}
</style>
	<link rel='stylesheet' id='crayon-css'  href='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/css/min/crayon.min.css?ver=_2.7.2_beta' type='text/css' media='all' />
<link rel='stylesheet' id='wp-block-library-css'  href='https://samnicholls.net/wp-includes/css/dist/block-library/style.min.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='child-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='lato-css'  href='//fonts.googleapis.com/css?family=Lato%3A400%2C700%2C400italic&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='slabo-css'  href='//fonts.googleapis.com/css?family=Slabo+27px&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='fa-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/font-awesome.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='rd-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-custom-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/custom.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='jetpack_css-css'  href='https://samnicholls.net/wp-content/plugins/jetpack/css/jetpack.css?ver=4.9' type='text/css' media='all' />
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery.min.js?ver=3.5.1' id='jquery-core-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery-migrate.min.js?ver=3.3.2' id='jquery-migrate-js'></script>
<script type='text/javascript' id='crayon_js-js-extra'>
/* <![CDATA[ */
var CrayonSyntaxSettings = {"version":"_2.7.2_beta","is_admin":"0","ajaxurl":"https:\/\/samnicholls.net\/wp-admin\/admin-ajax.php","prefix":"crayon-","setting":"crayon-setting","selected":"crayon-setting-selected","changed":"crayon-setting-changed","special":"crayon-setting-special","orig_value":"data-orig-value","debug":""};
var CrayonSyntaxStrings = {"copy":"Press %s to Copy, %s to Paste","minimize":"Click To Expand Code"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/js/min/crayon.min.js?ver=_2.7.2_beta' id='crayon_js-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/fd-footnotes/fdfootnotes.js?ver=1.34' id='fdfootnote_script-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/bootstrap.js?ver=5.7.5' id='bootstrap-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/custom.js?ver=5.7.5' id='js-custom-js'></script>
<link rel="https://api.w.org/" href="https://samnicholls.net/wp-json/" /><link rel="alternate" type="application/json" href="https://samnicholls.net/wp-json/wp/v2/posts/1581" /><link rel="EditURI" type="application/rsd+xml" title="RSD" href="https://samnicholls.net/xmlrpc.php?rsd" />
<link rel="wlwmanifest" type="application/wlwmanifest+xml" href="https://samnicholls.net/wp-includes/wlwmanifest.xml" /> 
<meta name="generator" content="WordPress 5.7.5" />
<link rel="canonical" href="https://samnicholls.net/2016/11/16/disorganised-disaster/" />
<link rel='shortlink' href='https://wp.me/p6RfP0-pv' />
<link rel="alternate" type="application/json+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F11%2F16%2Fdisorganised-disaster%2F" />
<link rel="alternate" type="text/xml+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F11%2F16%2Fdisorganised-disaster%2F&#038;format=xml" />
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-62206786-1', 'auto');
  ga('send', 'pageview');

</script>


<link rel='dns-prefetch' href='//v0.wordpress.com'>
<style type='text/css'>img#wpstats{display:none}</style>
<!-- Jetpack Open Graph Tags -->
<meta property="og:type" content="article" />
<meta property="og:title" content="Bioinformatics is a disorganised disaster and I am too. So I made a shell." />
<meta property="og:url" content="https://samnicholls.net/2016/11/16/disorganised-disaster/" />
<meta property="og:description" content="I am not a very organised bioinformatician, and despite your best efforts, you probably aren&#8217;t either. I&#8217;ve written a tool called `chitin` that tries to keep me, and bioinformatics moreâ€¦" />
<meta property="article:published_time" content="2016-11-16T17:50:59+00:00" />
<meta property="article:modified_time" content="2016-11-16T17:51:42+00:00" />
<meta property="og:site_name" content="Samposium" />
<meta property="og:image" content="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-16-11-11.png" />
<meta property="og:image:width" content="959" />
<meta property="og:image:height" content="982" />
<meta property="og:locale" content="en_GB" />
<meta name="twitter:card" content="summary" />
</head>

<body class="post-template-default single single-post postid-1581 single-format-standard">


        
<header id="main-header">
    <nav class="navbar-default navbar-static-top navbar" role="navigation">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#rd-collapse">
                    <span class="fa fa-bars"></span>
                    <span class="sr-only screen-reader-text">Toggle navigation</span>
                </button>
                <a class="navbar-brand" href=" https://samnicholls.net/">Samposium</a>            </div>
            
            <div id="rd-collapse" class="collapse navbar-collapse">
            
            <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body class="post-template-default single single-post postid-1581 single-format-standard">

<div id="main-menu" class="navbar-nav nav"><ul>
<li class="page_item page-item-8"><a href="https://samnicholls.net/about/">About Me</a></li>
<li class="page_item page-item-11"><a href="https://samnicholls.net/talks/">Talks</a></li>
<li class="page_item page-item-15"><a href="https://samnicholls.net/things-i-have-learned/">Things I Have Learned</a></li>
</ul></div></body></html>
        </div>

        </div>
    </nav>
</header>

<div class="container">

    <div class="row"><!-- Content -->
<section class="col-md-8 content-container">
                          <article class="card post hentry post-1581 type-post status-publish format-standard category-bioinf category-chitin tag-archive tag-bash tag-bioinformatics tag-chitin tag-database tag-disorganised tag-experiment tag-lab tag-labbook tag-organisation tag-pipeline tag-pitfalls tag-project tag-protocol tag-record tag-script tag-shell tag-sqlite tag-technique" id="post-1581">
                <header class="entry-header">
                        <h1 class="entry-title">Bioinformatics is a disorganised disaster and I am too. So I made a shell.</h1>                </header><!-- .entry-header -->

                <div class="meta-well">
                    <span style="margin-right:15px;" class=""><span class="fa fa-user"></span>&nbsp;&nbsp;&nbsp;<span class="author vcard"><a class="url fn n" href="https://samnicholls.net/about/">Sam Nicholls</a></span></span>

                    <span style="margin-right:15px;" class="entry-date published updated post-date"><span class="fa fa-clock-o"></span>&nbsp;&nbsp; <time class="entry-date published" datetime="2016-11-16T17:50:59+00:00">16th November 2016</time><time class="updated" datetime="2016-11-16T17:51:42+00:00">16th November 2016</time></span>                    <span style="margin-right:15px;" class=""><span class="fa fa-comments-o"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/2016/11/16/disorganised-disaster/#respond"><span class="dsq-postid" data-dsqidentifier="1581 https://samnicholls.net/?p=1581">No Comments yet</span></a></span>
                    <span class=""><span class="fa fa-folder-open"></span>&nbsp;&nbsp;&nbsp;<a href="https://samnicholls.net/category/bioinf/" rel="category tag">Bioinformatics</a>, <a href="https://samnicholls.net/category/bioinf/tools/chitin/" rel="category tag">chitin</a></span>
                </div>

                <div class="entry-content"><div class="note">If you don&#8217;t want to hear me wax lyrical about how disorganised I am, you can <a href="#chitin">skip ahead</a> to where I tell you about how great the pseudo-shell that I made and named <code>chitin</code> is.</div>
<p>Back in 2014, about half way through my undergraduate dissertation (<em><a href="https://github.com/samstudio8/frontier-dissertation">Application of Machine Learning Techniques to Next Generation Sequencing Quality Control</a></em>), I made an unsettling discovery.</p>
<p>I am disorganised.</p>
<p>The discovery was made after my supervisor asked a few interesting questions regarding some of my earlier discarded analyses. When I returned to the data to try and answer those questions, I found I simply could not regenerate the results. Despite the fact that both the code and each &#8220;experiment&#8221; were tracked by a <code>git</code> repository and I&#8217;d written my programs to output (what I thought to be) reasonable logs, I <em>still</em> could not reproduce my science. It could have been anything: an ad-hoc, temporary tweak to a harness script, a bug fix in the code itself masking a result, or any number of other possible untracked changes to the inputs or program parameters. In general, it was clear that I had failed to collect all pertinent metadata for <em>an experiment</em>.</p>
<p>Whilst it perhaps sounds like I was guilty of negligent book-keeping, it really wasn&#8217;t for lack of trying. Yet when dealing with many interesting questions at once, it&#8217;s so easy to make ad-hoc changes, or perform undocumented command line based munging of input data, or accidentally run a new experiment that clobbers something. Occasionally, one just forgets to make a note of something, or assumes a change is temporary but for one reason or another, the change becomes permanent without explanation. These subtle pipeline alterations are easily made all the time, and can silently invalidate swathes of results generated before (and/or after) them.</p>
<p>Ultimately, for the purpose of reproducibility, almost everything (copies of inputs, outputs, logs, configurations) was dumped and <code>tar</code>&#8216;d for each experiment. But this approach brought problems of its own: just tabulating results was difficult in its own right. In the end, I was pleased with that dissertation, but a small part of me still hurts when I think back to the problem of archiving and analysing those result sets.</p>
<p>It was a nightmare, and I promised it would never happen again.</p>
<p>Except it has.</p>
<h2>A relapse of disorganisation</h2>
<p>Two years later and I&#8217;ve continued to be capable of convincing a committee to allow me to progress towards adding the title of doctor to my bank account. As part of this quest, recently I was inspecting the results of a harness script responsible for <a href="https://samnicholls.net/2016/10/24/status-sep16/">generating trivial haplotypes, corresponding reads and attempting to recover them using <code>Gretel</code></a>. &#8220;Very interesting, but what will happen if I change the simulated read size&#8221;, I pondered; shortly before making an ad-hoc change to the harness script and inadvertently destroying the integrity of the results I had just finished inspecting by clobbering the input alignment file used as a parameter to <code>Gretel</code>.</p>
<p>Argh, not again.</p>
<h3>Why is this hard?</h3>
<p>Consider <code>Gretel</code>: she&#8217;s not just a simple standalone tool that one can execute to rescue haplotypes from the metagenome. One must go through the motions of pushing their raw reads through some form of pipeline (pictured below) to generate an <strong>alignment</strong> (to essentially give a co-ordinate system to those reads) and discover <strong>the variants</strong> (the positions in that co-ordinate system that relate to polymorphisms on reads) that form the required inputs for the recovery algorithm first.</p>
<p>This is problematic for one who wishes to be aware of the providence of all outputs of <code>Gretel</code>, as those outputs depend not only on the immediate inputs (the alignment and called variants), but the entirety of the pipeline that produced them. Thus we must capture as much information as possible regarding all of the steps that occur from the moment the raw reads hit the disk, up to <code>Gretel</code> finishing with extracted haplotypes.</p>
<p>But as I described in <a href="https://samnicholls.net/2016/10/24/status-sep16/">my last status report</a>, these tools are themselves non-trivial. <code>bowtie2</code> has more switches than an average spaceship, and its output depends on its complex set of parameters and inputs (that also have dependencies on previous commands), too.</p>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/11/IMG_20161110_103257.jpg"><img src="https://samnicholls.net/wp-content/uploads/2016/11/IMG_20161110_103257-1024x195.jpg" alt="img_20161110_103257" width="100%" class="alignnone size-large wp-image-1630" srcset="https://samnicholls.net/wp-content/uploads/2016/11/IMG_20161110_103257-1024x195.jpg 1024w, https://samnicholls.net/wp-content/uploads/2016/11/IMG_20161110_103257-300x57.jpg 300w, https://samnicholls.net/wp-content/uploads/2016/11/IMG_20161110_103257-768x146.jpg 768w" sizes="(max-width: 1024px) 100vw, 1024px" /></a></p>
<p><code>bash</code> scripts are all well and good for keeping track of a series of commands that yield the result of an experiment, and one can create a nice new directory in which to place such a result at the end &#8211; along with any log files and a copy of the harness script itself for good measure. But what happens when future experiments use different pipeline components, with different parameters, or we alter the generation of log files to make way for other metadata? What&#8217;s a good directory naming strategy for archiving results anyway? What if parts (or even all of the) analysis are ad-hoc and we are left to reconstruct the history? How many times have you made a manual edit to a malformed file, or had to look up exactly what combination of <code>sed</code>, <code>awk</code> and <code>grep</code> munging you did that one time?</p>
<p>One would have expected me to have learned my lesson by now, but I think meticulous digital lab book-keeping is just not that easy.</p>
<h3>What does organisation even mean anyway?</h3>
<p>I think the problem is perhaps exacerbated by conflating the meaning of &#8220;organisation&#8221;. There are a few somewhat different, but ultimately overlapping problems here:</p>
<ul>
<li><strong>How to keep track of how files are created</strong><br />
 What command created file <code>foo</code>? What were the parameters? When was it executed, by whom?</li>
<li><strong>Be aware of the role that each file plays in your pipeline</strong><br />
What commands go on to use file <code>foo</code>? Is it still needed?</li>
<li><strong>Assure the ongoing integrity of past and future results</strong><br />
Does this alignment have reads? Is that FASTA index up to date?<br />
Are we about to clobber shared inputs (large BAMS, references) that results depend on?</li>
<li><strong>Archiving results in a sensible fashion for future recall and comparison</strong><br />
How can we make it easy to find and analyse results in future?</li>
</ul>
<p>Indeed, my previous attempts at organisation address some <strong>but not all</strong> of these points, which is likely the source of my bad feeling. Keeping hold of <code>bash</code> scripts can help me determine how files are created, and the role those files go on to play in the pipeline; but results are merely dumped in a directory. Such directories are created with good intent, and named something that was likely useful and meaningful at the time. Unfortunately, I find that these directories become less and less useful as archive labels as time goes on&#8230;  For example, what the fuck is <code>../5-virus-mix/2016-10-11__ref896__reg2084-5083__sd100/</code><sup id="fnref-1581-1"><a href="#fn-1581-1" rel="footnote">1</a></sup>?</p>
<p>This approach also had no way to assure the current and future integrity of my results. Last month I had an issue with <code>Gretel</code> outputting bizarrely formatted haplotype <code>FASTA</code>s. After chasing my tail trying to find a bug in my <code>FASTA</code> I/O handling, I discovered this was actually caused by an out of date FASTA index (<code>.fai</code>) on the master reference. At some point I&#8217;d exchanged one FASTA for another, assuming that the index would be regenerated automatically. It wasn&#8217;t. Thus the integrity of experiments using that combination of FASTA+index was damaged. Additionally, the integrity of the results generated using the old FASTA were now also damaged: I&#8217;d clobbered the old master input.</p>
<p>There is a clear need to keep better metadata for files, executed commands and results, beyond just tracking everything with <code>git</code>. We need a better way to document the changes a command makes in the file system, and a mechanism to better assure integrity. Finally we need a method to archive experimental results in a more friendly way than a time-sensitive graveyard of timestamps, acronyms and abbreviations.</p>
<p>So I&#8217;ve taken it upon myself to get distracted from my PhD to embark on a new adventure to save myself from ruining my PhD<sup id="fnref-1581-2"><a href="#fn-1581-2" rel="footnote">2</a></sup>, and fix bioinformatics for everyone.</p>
<h2>Approaches for automated command collection</h2>
<p>Taking the number of post-its attached to my computer and my sporadically used notebooks as evidence enough to outright skip over the suggestion of a paper based solution to these problems, I see two schools of thought for capturing commands and metadata computationally:</p>
<ul>
<li><strong>Intrusive, but data is structured with perfect recall</strong><br />
A method whereby users must execute commands via some sort of wrapper. All commands must have some form of template that describes inputs, parameters and outputs. The wrapper then &#8220;fills in&#8221; the options and dispatches the command on the user&#8217;s behalf. All captured metadata has uniform structure and nicely avoids the need to attempt to parse user input. Command reconstruction is perfect but usage is arguably clunky.</li>
<li><strong>Unobtrusive, best-effort data collection</strong><br />
A daemon-like tool that attempts to collect executed commands from the user&#8217;s shell and monitor directories for file activity. Parsing command parameters and inputs is done in a naive best-effort scenario. The context of parsed commands and parameters is unknown; we don&#8217;t know what a particular command does, and cannot immediately discern between inputs, outputs, flags and arguments. But, despite the lack of structured data, the user does not notice our presence.</li>
</ul>
<p>There is a trade-off between usability and data quality here. If we sit between a user and all of their commands, offering a uniform interface to execute any piece of software, we can obtain perfectly structured information and are <strong>explicitly aware</strong> of parameter selections and the paths of all inputs and desired outputs. We know exactly where to monitor for file system changes, and can offer user interfaces that not only merely enumerate command executions, but offer searching and filtering capabilities based on captured parameters: &#8220;Show me assemblies that used a k-mer size of 31&#8221;.</p>
<p>But we must ask ourselves, how much is that fine-grained data worth to us? Is exchanging our ability to execute commands ourselves worth the perfectly structured data we can get via the wrapper? How much of those parameters are actually useful? Will I ever need to find all my <code>bowtie2</code> alignments that used 16 threads? There are other concerns here too: templates that define a job specification must be maintained. Someone must be responsible for adding new (or removing old) parameters to these templates when tools are updated. What if somebody happens to misconfigure such a template? More advanced users may be frustrated at being unable to merely execute their job on the command line. Less advanced users could be upset that they can&#8217;t just copy and paste commands from the manual or biostars. What about smaller jobs? Must one really define a command template to run trivial tools like <code>awk</code>, <code>sed</code>, <code>tail</code>, or <code>samtools sort</code> through the wrapper?</p>
<p>It turns out I know the answer to this already: the trade-off is not worth it.</p>
<h3>Intrusive wrappers don&#8217;t work: a sidenote on <code>sunblock</code></h3>
<p>Without wanting to bloat this post unnecessarily, I want to briefly discuss a tool I&#8217;ve written previously, but first I must set the scene<sup id="fnref-1581-3"><a href="#fn-1581-3" rel="footnote">3</a></sup>.</p>
<p>Within weeks of starting my PhD, I made a <a href="https://samnicholls.net/2015/02/17/sun-grid-engine/">computational enemy in the form of <em>Sun Grid Engine</em></a>: the scheduler software responsible for queuing, dispatching, executing and reporting on jobs submitted to the institute&#8217;s cluster. I rapidly became frustrated with having an unorganised collection of job scripts, with ad-hoc edits that meant I could no longer re-run a job previously executed with the same submission script (does this problem sound familiar?). In particular, I was upset with the state of the tools provided by SGE for reporting on the status of jobs.</p>
<p>To cheer myself up, I authored <a href="https://github.com/SamStudio8/sunblock">a tool called <code>sunblock</code></a>, with the goal of never having to look at any component of Sun Grid Engine directly ever again. I was successful in my endeavour and to this day continue to use the tool on the occasion where I need to use the cluster.</p>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-16-11-11.png"><img src="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-16-11-11.png" alt="screenshot-from-2016-11-16-16-11-11" width="60%" class="alignnone size-full wp-image-1685" srcset="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-16-11-11.png 959w, https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-16-11-11-293x300.png 293w, https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-16-11-11-768x786.png 768w" sizes="(max-width: 959px) 100vw, 959px" /></a></p>
<p>However, as hypothesised above, <code>sunblock</code> does indeed require an explicit description of an interface for any job that one would wish to submit to the cluster, and it does prevent users from just pasting commands into their terminal. This all-encompassing wrapping feature; that allows us to capture the best, structured information on every job, is also the tool&#8217;s complete downfall. Despite the useful information that could be extracted using <code>sunblock</code> (there is <a href="https://sunblock.ironowl.io/">even a shiny <code>sunblock</code> web interface</a>), its ability to automatically re-run jobs and the superior reporting on job progress compared to SGE alone, was still not enough to get user traction in our institute.</p>
<p>For the same reason that I think more in-the-know bioinformaticians don&#8217;t want to use <code>Galaxy</code>, <code>sunblock</code> failed: because <strong>it gets in the way</strong>.</p>
<p><a name="chitin"></a></p>
<h2>Introducing <code>chitin</code>: an awful shell for awful bioinformaticians</h2>
<p>Taking what I learned from my experimentation with <code>sunblock</code> on-board, I elected to take the less intrusive, best-effort route to collecting user commands and file system changes. Thus I introduce <code>chitin</code>: a Python based tool that (somewhat)-unobtrusively wraps your system shell, to keep track of commands and file manipulations to address the problem of not knowing how any of the files in your ridiculously complicated bioinformatics pipeline came to be.</p>
<p>I initially began the project with a view to create a digital lab book manager. I envisaged offering a command line tool with several subcommands, one of which could take a command for execution. However as soon as I tried out my prototype and found myself prepending the majority of my commands with <code>lab execute</code>, I wondered whether I could do better. What if I just wrapped the system shell and captured all entered commands? This might seem a rather dumb and long-about way of getting one&#8217;s command history, but consider this: if we wrap the system shell as a means to capture all the input, we are also in a position to capture the output for clever things, too. Imagine a shell that could parse the <code>stdout</code> for useful metadata to tag files with&#8230;</p>
<p>I liked what I was imagining, and so despite my best efforts to get even just one person to convince me otherwise; I wrote my own pseudo-shell.</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">WHAT&#39;S THAT? A SHELL WITH BUILT IN FUNCTIONS FOR HOW YOUR FILES HAPPENED AND WHAT YOU NEED TO REPEAT TO GET TO A GIVEN FILE? WHY YES IT IS <a href="https://t.co/h87pzptq1E">pic.twitter.com/h87pzptq1E</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/798347239634702340">November 15, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p><code>chitin</code> <strong>is already</strong> able to track executed commands that yield changes to the file system. For each file in the <code>chitin</code> tree, there is a full modification history. Better yet, you can ask what series of commands need to be executed in order to recreate a particular file in your workflow. It&#8217;s also possible to tag files with potentially useful metadata, and so <code>chitin</code> takes advantage of this by adding the runtime<sup id="fnref-1581-4"><a href="#fn-1581-4" rel="footnote">4</a></sup>, and current user to all executed commands for you.</p>
<p>Additionally, I&#8217;ve tried to find my own middle ground between the <code>sunblock</code>-esque configurations that yielded superior metadata, and not getting in the way of our users too much. So one may <strong>optionally</strong> specify handlers that can be applied to detected commands, and captured <code>stdout</code>/<code>stderr</code>. For example, thanks to my <code>bowtie2</code> configuration, <code>chitin</code> tags my <code>out.sam</code> files with the overall alignment rate (and a few targeted parameters of interest), <strong>automatically</strong>.</p>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-21-30.png"><img src="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-21-30.png" alt="screenshot-from-2016-11-16-17-21-30" width="80%" class="alignnone size-full wp-image-1707" srcset="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-21-30.png 875w, https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-21-30-300x104.png 300w, https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-21-30-768x266.png 768w" sizes="(max-width: 875px) 100vw, 875px" /></a></p>
<p><code>chitin</code> also allows you to specify handlers for particular file formats to be applied to files as they are encountered. My environment, for example, is set-up to count the number of reads inside a BAM, and associate that metadata with that version of the file:</p>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-30-55.png"><img src="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-30-55.png" alt="screenshot-from-2016-11-16-17-30-55" width="80%" class="alignnone size-full wp-image-1711" srcset="https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-30-55.png 875w, https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-30-55-300x56.png 300w, https://samnicholls.net/wp-content/uploads/2016/11/Screenshot-from-2016-11-16-17-30-55-768x142.png 768w" sizes="(max-width: 875px) 100vw, 875px" /></a></p>
<p>In this vein, we are in a nice position to check on the status of files before and after a command is executed. To address some of my integrity woes, <code>chitin</code> allows you to define integrity handlers for particular file formats too. Thus my environment warns me if a BAM has 0 reads, is missing an index, or has an index older than itself. Similarly, an empty VCF raises a warning, as does an out of date FASTA index. Coming shortly will be additional checks for whether you are about to clobber a file that is depended on by other files in your workflow. Kinda cool, even if I do say so myself.</p>
<h2>Conclusion</h2>
<p>Perhaps I&#8217;m trying to solve a problem of my own creation. Yet from a few conversations I&#8217;ve had with folks in my lab, and frankly, anyone I could get to listen to me for five minutes about managing bioinformatics pipelines, there seems to be sympathy to my cause. I&#8217;m not entirely convinced myself that a &#8220;shell&#8221; is the correct solution here, but it does seem to place us in the best position to get commands entered by the user, with the added bonus of getting <code>stdout</code> to parse for free. Though, judging by the flurry of Twitter activity on my dramatically posted <code>chitin</code> screenshots lately, I suspect I am not so alone in my disorganisation and there are at least a handful of bioinformaticians out there who think a shell isn&#8217;t the most terrible solution to this either. Perhaps I just need to be more of a wet-lab biologist.</p>
<p>Either way, I genuinely think there&#8217;s a lot of room to do cool stuff here, and to my surprise, I&#8217;m genuinely finding <code>chitin</code> quite useful already. If you&#8217;d like to try it out, <a href="https://github.com/SamStudio8/chitin">the source for <code>chitin</code> is open and free on GitHub</a>. Please don&#8217;t expect too much in the way of stability, though.</p>
<hr />
<h1>tl;dr</h1>
<ul>
<li>A definition of &#8220;being organised&#8221; for science and experimentation is hard to pin down</li>
<li>But independent of such a definition, I am terminally disorganised</li>
<li>Seriously what the fuck is <code>../5-virus-mix/2016-10-11__ref896__reg2084-5083__sd100</code><sup id="fnref2:1581-1"><a href="#fn-1581-1" rel="footnote">1</a></sup></li>
<li>I think command wrappers and platforms like <code>Galaxy</code> get in the way of things too much</li>
<li>I wrote a &#8220;shell&#8221; to try and compensate for this</li>
<li><a href="https://github.com/SamStudio8/chitin">Now I have a shell, it is called <code>chitin</code></a></li>
</ul>
<div class="footnotes">
<hr />
<ol>
<li id="fn-1581-1">
This is a genuine directory in my file system, created about a month ago. It contains results for a run of <code>Gretel</code> against the <code>pol</code> gene on the <code>HIV</code> genome (<code>2084-5083</code>). Off the top of my head, I cannot recall what <code>sd100</code> is, or why <code>reg</code> appears before the base positions. I honestly tried.&#160;<a href="#fnref-1581-1" rev="footnote">&#8617;</a> <a href="#fnref2:1581-1" rev="footnote">&#8617;</a>
</li>
<li id="fn-1581-2">
Because more things that are not my actual PhD is just what my PhD needs.&#160;<a href="#fnref-1581-2" rev="footnote">&#8617;</a>
</li>
<li id="fn-1581-3">
If it helps you, imagine some soft jazz playing to the sound of rain while I talk about this gruffly in the dark with a cigarette poking out of my mouth. Oh, and everything is in black and white. It&#8217;s <em>bioinformatique noir</em>.&#160;<a href="#fnref-1581-3" rev="footnote">&#8617;</a>
</li>
<li id="fn-1581-4">
I&#8217;m quite pleased with this one, because I pretty much always forget to <code>time</code> how long my assemblies and alignments take.&#160;<a href="#fnref-1581-4" rev="footnote">&#8617;</a>
</li>
</ol>
</div>
</div>
                
                        <div class="tag-well"><span><span class="fa fa-tags"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/tag/archive/" rel="tag">archive</a>,&nbsp; <a href="https://samnicholls.net/tag/bash/" rel="tag">bash</a>,&nbsp; <a href="https://samnicholls.net/tag/bioinformatics/" rel="tag">bioinformatics</a>,&nbsp; <a href="https://samnicholls.net/tag/chitin/" rel="tag">chitin</a>,&nbsp; <a href="https://samnicholls.net/tag/database/" rel="tag">database</a>,&nbsp; <a href="https://samnicholls.net/tag/disorganised/" rel="tag">disorganised</a>,&nbsp; <a href="https://samnicholls.net/tag/experiment/" rel="tag">experiment</a>,&nbsp; <a href="https://samnicholls.net/tag/lab/" rel="tag">lab</a>,&nbsp; <a href="https://samnicholls.net/tag/labbook/" rel="tag">labbook</a>,&nbsp; <a href="https://samnicholls.net/tag/organisation/" rel="tag">organisation</a>,&nbsp; <a href="https://samnicholls.net/tag/pipeline/" rel="tag">pipeline</a>,&nbsp; <a href="https://samnicholls.net/tag/pitfalls/" rel="tag">pitfalls</a>,&nbsp; <a href="https://samnicholls.net/tag/project/" rel="tag">project</a>,&nbsp; <a href="https://samnicholls.net/tag/protocol/" rel="tag">protocol</a>,&nbsp; <a href="https://samnicholls.net/tag/record/" rel="tag">record</a>,&nbsp; <a href="https://samnicholls.net/tag/script/" rel="tag">script</a>,&nbsp; <a href="https://samnicholls.net/tag/shell/" rel="tag">shell</a>,&nbsp; <a href="https://samnicholls.net/tag/sqlite/" rel="tag">sqlite</a>,&nbsp; <a href="https://samnicholls.net/tag/technique/" rel="tag">technique</a></span></div>
                
                <nav class="further-reading">
	<div class="previous">
		<span>Previous Post</span>
		<a href="https://samnicholls.net/2016/11/07/lightstick-p1/">Controlling a NeoPixel LED Strip and SD Card with a &lt;strike&gt;Arduino&lt;/strike&gt; ESP8266: My Lightstick Prototype</a>
	</div>
	<div class="next">
		<span>Next Post</span>
		<a href="https://samnicholls.net/2016/12/19/status-nov16-p1/">Status Report: November 2016 (Part I): Triviomes, Treeviomes &#038; Fuck Everything</a>
	</div>
</nav>            </article>
                                <div class="card">

<div id="disqus_thread"></div>
                </div>
    
</section>

        <!-- Sidebar -->
        <aside class="col-md-4 sidebar-container">
		            <div id="widget-area" class="widget-area" role="complementary">
                <div id="custom_html-3" class="widget_text card widget widget_custom_html"><div class="textwidget custom-html-widget"><img src="https://pbs.twimg.com/profile_images/1044255545610436609/wt33ZbtU_400x400.jpg" height=120 style="margin-top: 20px; margin-right:10px; float: left;" alt="Sam Nicholls" />
<div style="padding-top:15px; line-height: normal"><b style="font-size: 1.2em">Sam Nicholls</b><br><span style="font-size: 0.9em;">Haplotype Wrangler at University of Birmingham.<br>I tell computers to do things, then I write about how it all went wrong.</span></div></div></div><div id="search-2" class="card widget widget_search"><form style="margin-top:15px" action="https://samnicholls.net/" id="searchform" method="get" role="form">
    <div class="form-group">
        <input type="text" class="form-control" name="s" id="s" placeholder="e.g. Search...">
    </div>
</form>
</div><div id="twitter_timeline-5" class="card widget widget_twitter_timeline"><a class="twitter-timeline" data-height="400" data-theme="light" data-link-color="#f96e5b" data-border-color="#e8e8e8" data-lang="EN" data-partner="jetpack" data-chrome="noheader nofooter noborders" href="https://twitter.com/samstudio8">My Tweets</a></div><div id="categories-5" class="card widget widget_categories"><h2>Categories</h2>
			<ul>
					<li class="cat-item cat-item-2"><a href="https://samnicholls.net/category/bioinf/">Bioinformatics</a>
<ul class='children'>
	<li class="cat-item cat-item-8"><a href="https://samnicholls.net/category/bioinf/au-phd/">AU-PhD</a>
</li>
	<li class="cat-item cat-item-3"><a href="https://samnicholls.net/category/bioinf/sanger-qc/" title="Sanger QC Project">Sanger-QC</a>
</li>
	<li class="cat-item cat-item-61"><a href="https://samnicholls.net/category/bioinf/tools/">Tools</a>
	<ul class='children'>
	<li class="cat-item cat-item-185"><a href="https://samnicholls.net/category/bioinf/tools/chitin/">chitin</a>
</li>
	</ul>
</li>
</ul>
</li>
	<li class="cat-item cat-item-80"><a href="https://samnicholls.net/category/devops/">Devops</a>
</li>
	<li class="cat-item cat-item-114"><a href="https://samnicholls.net/category/event/">Event</a>
</li>
	<li class="cat-item cat-item-21"><a href="https://samnicholls.net/category/gadgets/">Gadgets</a>
</li>
	<li class="cat-item cat-item-176"><a href="https://samnicholls.net/category/hardware/">Hardware</a>
<ul class='children'>
	<li class="cat-item cat-item-223"><a href="https://samnicholls.net/category/hardware/beehive-monitor/">Beehive Monitor</a>
</li>
	<li class="cat-item cat-item-178"><a href="https://samnicholls.net/category/hardware/esp8266/">ESP8266</a>
</li>
	<li class="cat-item cat-item-236"><a href="https://samnicholls.net/category/hardware/lego-sequencer/">Lego Sequencer</a>
</li>
	<li class="cat-item cat-item-177"><a href="https://samnicholls.net/category/hardware/lightstick/">Lightstick</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-151"><a href="https://samnicholls.net/category/lab/">Lab</a>
<ul class='children'>
	<li class="cat-item cat-item-150"><a href="https://samnicholls.net/category/lab/protocol-guides/">Protocol Guides</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-5"><a href="https://samnicholls.net/category/meta/">Meta</a>
</li>
	<li class="cat-item cat-item-62"><a href="https://samnicholls.net/category/mysteries/">Mysteries</a>
</li>
	<li class="cat-item cat-item-22"><a href="https://samnicholls.net/category/personal/">Personal</a>
</li>
	<li class="cat-item cat-item-4"><a href="https://samnicholls.net/category/photo/">Photography</a>
</li>
	<li class="cat-item cat-item-143"><a href="https://samnicholls.net/category/programming/">Programming</a>
<ul class='children'>
	<li class="cat-item cat-item-145"><a href="https://samnicholls.net/category/programming/documentation/">Documentation</a>
</li>
	<li class="cat-item cat-item-144"><a href="https://samnicholls.net/category/programming/python/">Python</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-48"><a href="https://samnicholls.net/category/status-report/">Status Report</a>
</li>
	<li class="cat-item cat-item-60"><a href="https://samnicholls.net/category/sysadm/">System Administration</a>
</li>
	<li class="cat-item cat-item-1"><a href="https://samnicholls.net/category/uncategorised/">Uncategorised</a>
</li>
			</ul>

			</div>            </div>
		        </aside>
    </div>

</div> <!-- /container -->

<footer id="main-footer" class="bg-primary">
	<div id="footer-widgets" class="container">
		    </div>
</footer>



	<div style="display:none">
	</div>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/comment-reply.min.js?ver=5.7.5' id='comment-reply-js'></script>
<script type='text/javascript' id='disqus_count-js-extra'>
/* <![CDATA[ */
var countVars = {"disqusShortname":"samposium-blog"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_count.js?ver=3.0.17' id='disqus_count-js'></script>
<script type='text/javascript' id='disqus_embed-js-extra'>
/* <![CDATA[ */
var embedVars = {"disqusConfig":{"integration":"wordpress 3.0.17"},"disqusIdentifier":"1581 https:\/\/samnicholls.net\/?p=1581","disqusShortname":"samposium-blog","disqusTitle":"Bioinformatics is a disorganised disaster and I am too. So I made a shell.","disqusUrl":"https:\/\/samnicholls.net\/2016\/11\/16\/disorganised-disaster\/","postId":"1581"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_embed.js?ver=3.0.17' id='disqus_embed-js'></script>
<script type='text/javascript' src='https://s0.wp.com/wp-content/js/devicepx-jetpack.js?ver=202335' id='devicepx-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/jetpack/_inc/twitter-timeline.js?ver=4.0.0' id='jetpack-twitter-timeline-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/wp-embed.min.js?ver=5.7.5' id='wp-embed-js'></script>
<script type='text/javascript' src='https://stats.wp.com/e-202335.js' async defer></script>
<script type='text/javascript'>
	_stq = window._stq || [];
	_stq.push([ 'view', {v:'ext',j:'1:4.9',blog:'101350222',post:'1581',tz:'1',srv:'samnicholls.net'} ]);
	_stq.push([ 'clickTrackerInit', '101350222', '1581' ]);
</script>

<script type="text/javascript">
(function() {
  if (navigator.userAgent.match(/IEMobile\/10\.0/)) {
    var msViewportStyle = document.createElement("style");
    msViewportStyle.appendChild(
      document.createTextNode("@-ms-viewport{width:auto!important}")
    );
    document.getElementsByTagName("head")[0].appendChild(msViewportStyle);
  }
})();

var collapsed = false;
jQuery( document ).ready(function($) {
    $('.navbar-toggle').click(function(){
        console.log(collapsed);
        if(collapsed) {
            $('.navbar-header > button').html('<i class="fa fa-bars"></i>');
        } else {
            $('.navbar-header > button').html('<i class="fa fa-times"></i>');
        }
        $('.navbar-collapse').toggle(200);
        collapsed = !collapsed;
    });
});
</script>
	
</body>

</html>
