<!DOCTYPE html>
<html lang="en-GB">

<head>
	<meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="pingback" href="https://samnicholls.net/xmlrpc.php">

	
	<title>Status Report: Jul Aug September 2016 &#8211; Samposium</title>
<meta name='robots' content='max-image-preview:large' />
<link rel='dns-prefetch' href='//s0.wp.com' />
<link rel='dns-prefetch' href='//fonts.googleapis.com' />
<link rel='dns-prefetch' href='//s.w.org' />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Feed" href="https://samnicholls.net/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Comments Feed" href="https://samnicholls.net/comments/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Status Report: Jul Aug September 2016 Comments Feed" href="https://samnicholls.net/2016/10/24/status-sep16/feed/" />
		<script type="text/javascript">
			window._wpemojiSettings = {"baseUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/72x72\/","ext":".png","svgUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/svg\/","svgExt":".svg","source":{"concatemoji":"https:\/\/samnicholls.net\/wp-includes\/js\/wp-emoji-release.min.js?ver=5.7.5"}};
			!function(e,a,t){var n,r,o,i=a.createElement("canvas"),p=i.getContext&&i.getContext("2d");function s(e,t){var a=String.fromCharCode;p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,e),0,0);e=i.toDataURL();return p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,t),0,0),e===i.toDataURL()}function c(e){var t=a.createElement("script");t.src=e,t.defer=t.type="text/javascript",a.getElementsByTagName("head")[0].appendChild(t)}for(o=Array("flag","emoji"),t.supports={everything:!0,everythingExceptFlag:!0},r=0;r<o.length;r++)t.supports[o[r]]=function(e){if(!p||!p.fillText)return!1;switch(p.textBaseline="top",p.font="600 32px Arial",e){case"flag":return s([127987,65039,8205,9895,65039],[127987,65039,8203,9895,65039])?!1:!s([55356,56826,55356,56819],[55356,56826,8203,55356,56819])&&!s([55356,57332,56128,56423,56128,56418,56128,56421,56128,56430,56128,56423,56128,56447],[55356,57332,8203,56128,56423,8203,56128,56418,8203,56128,56421,8203,56128,56430,8203,56128,56423,8203,56128,56447]);case"emoji":return!s([55357,56424,8205,55356,57212],[55357,56424,8203,55356,57212])}return!1}(o[r]),t.supports.everything=t.supports.everything&&t.supports[o[r]],"flag"!==o[r]&&(t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&t.supports[o[r]]);t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&!t.supports.flag,t.DOMReady=!1,t.readyCallback=function(){t.DOMReady=!0},t.supports.everything||(n=function(){t.readyCallback()},a.addEventListener?(a.addEventListener("DOMContentLoaded",n,!1),e.addEventListener("load",n,!1)):(e.attachEvent("onload",n),a.attachEvent("onreadystatechange",function(){"complete"===a.readyState&&t.readyCallback()})),(n=t.source||{}).concatemoji?c(n.concatemoji):n.wpemoji&&n.twemoji&&(c(n.twemoji),c(n.wpemoji)))}(window,document,window._wpemojiSettings);
		</script>
		<style type="text/css">
img.wp-smiley,
img.emoji {
	display: inline !important;
	border: none !important;
	box-shadow: none !important;
	height: 1em !important;
	width: 1em !important;
	margin: 0 .07em !important;
	vertical-align: -0.1em !important;
	background: none !important;
	padding: 0 !important;
}
</style>
	<link rel='stylesheet' id='crayon-css'  href='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/css/min/crayon.min.css?ver=_2.7.2_beta' type='text/css' media='all' />
<link rel='stylesheet' id='wp-block-library-css'  href='https://samnicholls.net/wp-includes/css/dist/block-library/style.min.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='child-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='lato-css'  href='//fonts.googleapis.com/css?family=Lato%3A400%2C700%2C400italic&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='slabo-css'  href='//fonts.googleapis.com/css?family=Slabo+27px&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='fa-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/font-awesome.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='rd-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-custom-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/custom.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='jetpack_css-css'  href='https://samnicholls.net/wp-content/plugins/jetpack/css/jetpack.css?ver=4.9' type='text/css' media='all' />
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery.min.js?ver=3.5.1' id='jquery-core-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery-migrate.min.js?ver=3.3.2' id='jquery-migrate-js'></script>
<script type='text/javascript' id='crayon_js-js-extra'>
/* <![CDATA[ */
var CrayonSyntaxSettings = {"version":"_2.7.2_beta","is_admin":"0","ajaxurl":"https:\/\/samnicholls.net\/wp-admin\/admin-ajax.php","prefix":"crayon-","setting":"crayon-setting","selected":"crayon-setting-selected","changed":"crayon-setting-changed","special":"crayon-setting-special","orig_value":"data-orig-value","debug":""};
var CrayonSyntaxStrings = {"copy":"Press %s to Copy, %s to Paste","minimize":"Click To Expand Code"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/js/min/crayon.min.js?ver=_2.7.2_beta' id='crayon_js-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/fd-footnotes/fdfootnotes.js?ver=1.34' id='fdfootnote_script-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/bootstrap.js?ver=5.7.5' id='bootstrap-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/custom.js?ver=5.7.5' id='js-custom-js'></script>
<link rel="https://api.w.org/" href="https://samnicholls.net/wp-json/" /><link rel="alternate" type="application/json" href="https://samnicholls.net/wp-json/wp/v2/posts/1201" /><link rel="EditURI" type="application/rsd+xml" title="RSD" href="https://samnicholls.net/xmlrpc.php?rsd" />
<link rel="wlwmanifest" type="application/wlwmanifest+xml" href="https://samnicholls.net/wp-includes/wlwmanifest.xml" /> 
<meta name="generator" content="WordPress 5.7.5" />
<link rel="canonical" href="https://samnicholls.net/2016/10/24/status-sep16/" />
<link rel='shortlink' href='https://wp.me/p6RfP0-jn' />
<link rel="alternate" type="application/json+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F10%2F24%2Fstatus-sep16%2F" />
<link rel="alternate" type="text/xml+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F10%2F24%2Fstatus-sep16%2F&#038;format=xml" />
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-62206786-1', 'auto');
  ga('send', 'pageview');

</script>


<link rel='dns-prefetch' href='//v0.wordpress.com'>
<style type='text/css'>img#wpstats{display:none}</style>
<!-- Jetpack Open Graph Tags -->
<meta property="og:type" content="article" />
<meta property="og:title" content="Status Report: Jul Aug September 2016" />
<meta property="og:url" content="https://samnicholls.net/2016/10/24/status-sep16/" />
<meta property="og:description" content="A status update on the apparent progress of my doctorate, featuring a name for my work, and a brief run through all of the things that are and are not broken as of September 2016." />
<meta property="article:published_time" content="2016-10-24T15:37:02+00:00" />
<meta property="article:modified_time" content="2016-10-24T15:54:28+00:00" />
<meta property="og:site_name" content="Samposium" />
<meta property="og:image" content="https://samnicholls.net/wp-content/uploads/2016/10/hansel2-1024x540.png" />
<meta property="og:locale" content="en_GB" />
<meta name="twitter:card" content="summary" />
</head>

<body class="post-template-default single single-post postid-1201 single-format-standard">


        
<header id="main-header">
    <nav class="navbar-default navbar-static-top navbar" role="navigation">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#rd-collapse">
                    <span class="fa fa-bars"></span>
                    <span class="sr-only screen-reader-text">Toggle navigation</span>
                </button>
                <a class="navbar-brand" href=" https://samnicholls.net/">Samposium</a>            </div>
            
            <div id="rd-collapse" class="collapse navbar-collapse">
            
            <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body class="post-template-default single single-post postid-1201 single-format-standard">

<div id="main-menu" class="navbar-nav nav"><ul>
<li class="page_item page-item-8"><a href="https://samnicholls.net/about/">About Me</a></li>
<li class="page_item page-item-11"><a href="https://samnicholls.net/talks/">Talks</a></li>
<li class="page_item page-item-15"><a href="https://samnicholls.net/things-i-have-learned/">Things I Have Learned</a></li>
</ul></div></body></html>
        </div>

        </div>
    </nav>
</header>

<div class="container">

    <div class="row"><!-- Content -->
<section class="col-md-8 content-container">
                          <article class="card post hentry post-1201 type-post status-publish format-standard category-au-phd category-status-report tag-metahaplome tag-paper tag-preprint tag-sitrep tag-status-report" id="post-1201">
                <header class="entry-header">
                        <h1 class="entry-title">Status Report: <strike>Jul</strike> <strike>Aug</strike> September 2016</h1>                </header><!-- .entry-header -->

                <div class="meta-well">
                    <span style="margin-right:15px;" class=""><span class="fa fa-user"></span>&nbsp;&nbsp;&nbsp;<span class="author vcard"><a class="url fn n" href="https://samnicholls.net/about/">Sam Nicholls</a></span></span>

                    <span style="margin-right:15px;" class="entry-date published updated post-date"><span class="fa fa-clock-o"></span>&nbsp;&nbsp; <time class="entry-date published" datetime="2016-10-24T16:37:02+01:00">24th October 2016</time><time class="updated" datetime="2016-10-24T16:54:28+01:00">24th October 2016</time></span>                    <span style="margin-right:15px;" class=""><span class="fa fa-comments-o"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/2016/10/24/status-sep16/#respond"><span class="dsq-postid" data-dsqidentifier="1201 https://samnicholls.net/?p=1201">No Comments yet</span></a></span>
                    <span class=""><span class="fa fa-folder-open"></span>&nbsp;&nbsp;&nbsp;<a href="https://samnicholls.net/category/bioinf/au-phd/" rel="category tag">AU-PhD</a>, <a href="https://samnicholls.net/category/status-report/" rel="category tag">Status Report</a></span>
                </div>

                <div class="entry-content"><p>A sufficient time had passed since my <a href="https://samnicholls.net/2016/06/12/status-may16/">previous report</a> such that there were both a number of things to report, and I crossed the required threshold of edginess to provide an update on the progress of the long list of things I am supposed to be doing in my quest to become Dr. Nicholls. I began a draft of this post shortly before I attended the European Conference on Computational Biology at the start of September. However at the end of the conference, I spontaneously decided to temporarily abort my responsibilities to bioinformatics, not return to Aberystwyth, electing to spend a few weeks traversing Eastern Europe instead. I imagine there will be more on this in a future post, but for now let&#8217;s focus on the long overdue insight into the work that I am supposed to be doing.</p>
<p>In our last installment, I finally managed to describe in words how the <a href="https://samnicholls.net/2016/03/01/status-feb16/"><strong>metahaplome</strong></a> is somewhat like a graph, but possesses properties that also make it not like a graph. The main outstanding issues were related to reweighting evidence in the metahaplome structure, the ongoing saga of generating sufficient data sets for evaluation and writing this all up so people believe us when we say it is possible. Of the many elephants in the room, the largest was my work still being described as <strong>the metahaplome</strong> or some flavour of <strong>Sam&#8217;s algorithm</strong>. It was time to try and conquer one of the more difficult problems in computer science: naming things.</p>
<h2>Introducing <code>Hansel</code> and <code>Gretel</code></h2>
<p>Harbouring a dislike for the apparent convention in bioinformatics that software names should be unimaginative<sup id="fnref-1201-1"><a href="#fn-1201-1" rel="footnote">1</a></sup> or awkwardly constructed acronyms, and spotting an opportunity to acquire a whimsical theme, I decided to follow in the footsteps of <a href="http://bioinformatics.oxfordjournals.org/content/32/13/2047"><strong>Goldilocks</strong></a>, and continue the fairy tale naming scheme; on the condition I could find a name that fit comfortably.</p>
<p>Unsurprisingly, this proved quite difficult. After some debate, the most suitable fairy tale name I could find was <code>Rumpelstiltskin</code>, for its &#8216;straw-to-gold&#8217; reference. I liked the idea of your straw-like reads being converted to golden haplotypes. However as I am not remotely capable of spelling the name without Google, and the link between the name and software feels a tad tenuous, I vetoed the option and forged forward with the paper, with the algorithm <code>Untitled</code>.</p>
<p>As I considered the logistics of packaging the implementation as it stood, I realised that I had essentially created both a novel data structure for storing the metahaplome, as well as an actual algorithm for utilising that information to recover haplotypes from a metagenome. At this point everything fell into place; a nice packaging solution and a fitting pair of names had resolved themselves. Thus, I introduce the <code>Hansel</code> data structure, and the <code>Gretel</code> algorithm; a framework for recovering real haplotypes from metagenomes.</p>
<h3>Hansel</h3>
<p><code>Hansel</code> is a Python package that houses our novel data structure. I describe it as a &#8220;graph-inspired data structure for determining likely chains of symbol sequences from crummy evidence&#8221;. Hansel is designed to work with counts of pairwise co-occurrences of symbols in space or time. For our purposes, those symbols are the chemical bases of DNA (or RNA, or perhaps even amino acids of proteins), and the &#8220;space&#8221; dimension is their position along some sequence.</p>
<div id="attachment_1278" style="width: 440px" class="wp-caption aligncenter"><a href="https://samnicholls.net/wp-content/uploads/2016/10/hansel6.png"><img aria-describedby="caption-attachment-1278" loading="lazy" class="wp-image-1278 size-large" src="https://samnicholls.net/wp-content/uploads/2016/10/hansel6-845x1024.png" width="430" height="521" srcset="https://samnicholls.net/wp-content/uploads/2016/10/hansel6-845x1024.png 845w, https://samnicholls.net/wp-content/uploads/2016/10/hansel6-248x300.png 248w, https://samnicholls.net/wp-content/uploads/2016/10/hansel6-768x931.png 768w" sizes="(max-width: 430px) 100vw, 430px" /></a><p id="caption-attachment-1278" class="wp-caption-text">Three corresponding representations, (top) aligned reads, (middle) the Hansel structure, (bottom) a graph that can be derived from the Hansel structure</p></div>
<p>We fill this structure with counts of the number of times we observe some pair of nucleotides, at some pair of positions on the same read. Hansel provides a <code>numpy</code> <code>ndarray</code> backed class that offers a user friendly API for operating on our data structure: including adding, adjusting and fetching the counts of pairwise co-occurrences of pairs of symbols in space and time, and making probabilistic queries on the likelihood of symbols occurring, given some sequence of observed symbols thus far.</p>
<h3>Gretel</h3>
<p><code>Gretel</code> is a Python package that now houses the actual algorithm that exploits the spun-out API offered by <code>Hansel</code> to attempt to recover haplotypes from the metagenome. <code>Gretel</code> provides a command line based tool that accepts a BAM of aligned reads (against a metagenomic pseudo-reference, typically the assembly) and a VCF of single nucleotide polymorphisms.</p>
<p><code>Gretel</code> populates the <code>Hansel</code> matrix with observations by parsing the reads in the BAM at the genomic positions described in the provided VCF. Once parsing the reads is complete, <code>Gretel</code> exploits the ability to traverse the <code>Hansel</code> structure like a graph, creating chains of nucleotides that represent the most likely sequence of SNPs that appear on reads that in turn align to some region of interest on the pseudo-reference.</p>
<p>At each node, the decision to move to the next node (<em>i.e.</em> nucleotide) is made by ranking each of the probabilities of the node on the end of each current outgoing edge appearing after some subset of the previously seen nodes (<em>i.e.</em> the current path). Thus both the availability and associated probabilities of future edges are dependent on not only the current or previous node, but the path itself.</p>
<div id="attachment_1274" style="width: 583px" class="wp-caption aligncenter"><a href="https://samnicholls.net/wp-content/uploads/2016/10/hansel2.png"><img aria-describedby="caption-attachment-1274" loading="lazy" class="wp-image-1274 size-large" src="https://samnicholls.net/wp-content/uploads/2016/10/hansel2-1024x540.png" alt="Pairwise conditionals between L last variants on the observed path, and each of the possible next variants are calculated and the best option (highest likelihood) is chosen" width="573" height="302" srcset="https://samnicholls.net/wp-content/uploads/2016/10/hansel2-1024x540.png 1024w, https://samnicholls.net/wp-content/uploads/2016/10/hansel2-300x158.png 300w, https://samnicholls.net/wp-content/uploads/2016/10/hansel2-768x405.png 768w" sizes="(max-width: 573px) 100vw, 573px" /></a><p id="caption-attachment-1274" class="wp-caption-text">Pairwise conditionals between L last variants on the observed path,<br /> and each of the possible next variants are calculated and the<br /> best option (highest likelihood) is chosen</p></div>
<p><code>Gretel</code> will construct a path in this way until a dummy sentinel representing the end of the graph is reached. After a path has been constructed from the dummy (or &#8220;sentinel&#8221;) start node to the end, observations in the <code>Hansel</code> structure are reweighted; to ensure that the same path is not returned again (as traversal is deterministic) and to allow <code>Gretel</code> to return the next most likely haplotype instead.</p>
<p><code>Gretel</code> repeatedly attempts to traverse the graph, each time returning the next most likely haplotype given the reweighted <code>Hansel</code> structure, until a node with no outgoing edges is encountered (<em>i.e.</em> all observations between two genomic positions have been culled by re-weighting).</p>
<h2>Going Public</h2>
<p>After surgically separating <code>Hansel</code> from <code>Gretel</code>, the codebases were finally in a state where I wasn&#8217;t sick at the thought of anyone looking at them. Thus <code>Hansel</code> and <code>Gretel</code> now have homes on Github: see <a href="https://github.com/samstudio8/hansel">github.com/samstudio8/hansel</a> and <a href="https://github.com/samstudio8/gretel">github.com/samstudio8/gretel</a>. In an attempt to not be an awful person, partial documentation now also exists on ReadTheDocs: see <a href="http://hansel.readthedocs.io/en/latest/">hansel.readthedocs.io</a> and <a href="http://gretel.readthedocs.io/en/latest/">gretel.readthedocs.io</a>. <code>Hansel</code> and <code>Gretel</code> are both open source and distributed under the MIT Licence.</p>
<h2>Reweighting</h2>
<p>For the time being, I am settled on the methodology for reweighting the <code>Hansel</code> matrix as described in <a href="https://samnicholls.net/2016/03/01/status-feb16/">February&#8217;s status report</a>, with the only recent difference being that it is <a href="https://samnicholls.net/2016/06/12/status-may16/">now implemented correctly</a>. Given a path through the metahaplome found by <code>Gretel</code> (that is, a predicted haplotype), we consider the marginal distribution of each of the selected variants that compose that path, and select the smallest (<em>i.e.</em> the least likely variant to be selected across all SNPs). That probability is then used as a ratio to reduce the observations in the <code>Hansel</code> data structure for all pairs of variants on the path.</p>
<p>As I have mentioned in the past, reweighting requires care: an overly aggressive methodology would dismiss similar looking paths (haplotypes) that have shared variants, but an overly cautious algorithm would limit exploration of the metahaplome structure and inhibit recovery of real haplotypes.</p>
<p>Testing so far has indicated that out of the methods I have proposed, this is the most robust, striking a happy balance between exploration-potential and accuracy. At this time I&#8217;m happy to leave it as is and work on something else, before I break it horribly.</p>
<h2>The &#8220;Lookback&#8221; Parameter</h2>
<p>As the calculation of edge likelihoods depends on <em>a subset of</em> the current path, a keen reader might wonder how such a subset is defined. Your typical bioinformatics software author would likely leave this as an exercise:<br />
<code>-l, --lookback     [mysteriously important non-optional integer that fundamentally changes how well this software works, correct selection is a dark art, never set to 0, good luck.]</code></p>
<p>Luckily for both bioinformatics and my <em>viva</em>, I refuse to inflict the field&#8217;s next <em>k-mer size</em>, and there exists a reasonable intuition for the selection of <code>L</code>: the number of nodes over which to <em>lookback</em> (from and including the head of the current path) when considering the next node (nucleotide of the sequence). My intuition is that sequenced read fragments cover some limited number of SNP sites. Thus there will be some <code>L</code> after which it is <em>typically</em> unlikely that pairs of SNPs will co-occur on the same read between the next variant node <code>i+1</code> and the already observed node <code>(i+1) - L</code>.</p>
<p>I say <em>typically</em> here because in actuality, SNPs will not be uniformly distributed<sup id="fnref-1201-2"><a href="#fn-1201-2" rel="footnote">2</a></sup> and as such the value of <code>L</code> according to this intuition is going to vary across the region of interest depending on both the read sizes, mate pair insert sizes, coverage and indeed the distribution of the SNPs themselves. We thus consider the average number of SNPs covered by all reads parsed from the BAM, and use this to set the value of <code>L</code>.</p>
<h2>Smoothing</h2>
<p>Smoothing is a component of <code>Hansel</code> that has flown under the radar in my previous discussions. This is likely because I have not given it a second thought since <a href="https://samnicholls.net/2015/08/04/belgian/">being Belgian last year</a>. Smoothing attempts to awkwardly sidestep two potential issues, namely <strong>overfitting</strong> and <strong><code>ZeroDivisionError</code></strong>.</p>
<p>For the former, we want to avoid scenarios where variant sites with very low read coverage (and thus few informative observations) are assumed to be fully representative of the true variation. In this case, smoothing effectively inserts additional observations that were not actually present in the provided reads to attempt to make up for potentially unseen observations.</p>
<p>For the latter case, consider the conditional probability of symbol <code>a</code> at position <code>i</code> occurring, given some symbol <code>b</code> at position <code>j</code>. This is defined (before smoothing is applied) by the <code>Hansel</code> framework as:</p>
<p><img src="//s0.wp.com/latex.php?latex=%5Cfrac%7B%5Ctext%7BReads+featuring%7D+%7Ea%7E+%5Ctext%7Bat%7D+%7Ei%7E+%5Ctext%7Band%7D+%7Eb%7E+%5Ctext%7Bat%7D+%7Ej%7D%7B%5Ctext%7BReads+spanning%7D+%7Ei%7E+%5Ctext%7Band+feature%7D+%7Eb%7E+%5Ctext%7Bat%7D+%7Ej%7D++&#038;bg=ffffff&#038;fg=000&#038;s=2" alt="&#92;frac{&#92;text{Reads featuring} ~a~ &#92;text{at} ~i~ &#92;text{and} ~b~ &#92;text{at} ~j}{&#92;text{Reads spanning} ~i~ &#92;text{and feature} ~b~ &#92;text{at} ~j}  " title="&#92;frac{&#92;text{Reads featuring} ~a~ &#92;text{at} ~i~ &#92;text{and} ~b~ &#92;text{at} ~j}{&#92;text{Reads spanning} ~i~ &#92;text{and feature} ~b~ &#92;text{at} ~j}  " class="latex" /></p>
<p>If one were to query the <code>Hansel</code> API with some selection of <code>i</code>, <code>b</code> and <code>j</code> such that there are no reads spanning <code>i</code> that feature symbol <code>b</code> at position <code>j</code>, a <code>ZeroDivisionError</code> will be raised. This is undesirable and cannot be circumvented by simply &#8220;catching&#8221; the error and returning 0, as the inclusion of a probability of 0 in a sequence of products renders the entire sequence of probabilities as 0, too.</p>
<p>The current smoothing method is merely &#8220;add one smoothing&#8221;, which modifies the above equation to artificially insert one (more) observation for each <em>possible</em> combination of symbols <code>a</code> and <code>b</code> between <code>i</code> and <code>j</code>. This avoids division by zero as there will always be at least one valid observation. However I suspect that to truly address the former problem, a more sophisticated solution is necessary.</p>
<p>Fortunately, it appears that in practice, for metagenomes with reasonable coverage, the first problem falls away and smoothing has a negligible effect on evaluation of edge probabilities. Despite this, the method of smoothing employed is admittedly naive and future work could potentially benefit from replacement. It should be noted that the influence of smoothing has the potential to become particularly pronounced after significant reweighting of the <code>Hansel</code> matrix (<em>i.e.</em> when very few observations remain).</p>
<h2>Evaluation</h2>
<p>Avid readers will be aware that evaluation of this method has been a persistent thorn in my side since the very beginning. There are <strong>no</strong> metagenomic test data sets that offer raw sequence reads, an assembly of those reads <strong>and</strong> a set of all expected (or indeed, even just some) haplotypes for a gene known to truly exist in that metagenome. Clearly this makes evaluation of a method for enumerating the most likely haplotypes of a particular gene from sequenced metagenomic reads somewhat difficult.</p>
<p>You might remember <a href="https://samnicholls.net/2016/06/12/status-may16/">from my last status report</a> that generating my own test data sets for real genes was both convoluted, and fraught with inconsistencies that made it difficult to determine whether unrecovered variants were a result of my approach, or an artefact of the data itself. I decided to take a step back and consider a simpler form of the problem, in the hope that I could construct an adequate testing framework for future development, and provide an empirical proof that the approach works (as well as a platform to investigate the conditions under which it doesn&#8217;t).</p>
<h3>Trivial Haplomes: <strong>Triviomes</strong></h3>
<p>To truly understand and diagnose the decisions my algorithm was making, I needed a source of reads that could be finely controlled and well-defined. I also needed the workflow to generate those reads to be free of uncontrolled external influences (<em>i.e.</em> reads dropped by alignment, difficult to predict SNP calling).</p>
<p>To accomplish this I created a script to construct <em>trivial haplomes</em>: sets of short, randomly generated haplotypes, each of the same fixed length. Every genomic position of such a trivial haplotype was considered to be a site of variation (<em>i.e.</em> a SNP). Tiny reads (of a configurable size range, set to 3-5bp for my work) are then constructed by sliding windows across each of the random haplotypes. Additional options exist to set a per-base error rate and &#8220;slip&#8221; the window (to reduce the quality of some read overlaps, decreasing the number of shared paired observations).</p>
<p>Although this appears to be grossly unrepresentative of the real problem at hand &#8212; what technology even generates reads of 3-5bp? Don&#8217;t forget, <code>Hansel</code> and <code>Gretel</code> are designed to work directly with SNPs. Aligned reads are only parsed at positions specified in the VCF (<em>i.e.</em> the list of called variants) and so real sequences collapse to a handful of SNPs anyway. The goal here is not so much about accurately emulating real reads with real variation and error, but to establish a framework for controlled performance testing under varying conditions (<em>e.g.</em> how do recovery rates change with respect to alignment rate, error rate, number of SNPs, haplotypes <em>etc</em>).</p>
<p>We must also isolate the generation of input files (alignments and variant lists) from external processes. That is, we cannot use established tools for read alignment and variant calling. Whilst the results of these processes are tractable and deterministic, they confound the testing of my triviomes due to their non-trivial behaviour. For example, my tiny reads have a known <em>correct</em> alignment: each read is yielded from some window with a defined start and end position on a given haplotype. However read aligners discard, clip and &#8220;incorrectly&#8221; align these tiny test reads<sup id="fnref-1201-3"><a href="#fn-1201-3" rel="footnote">3</a></sup>. My reads are no longer under my direct control.</p>
<p>In the same fashion, despite my intention that every genomic position of my trivial haplome is a SNP, established variant callers and their diploid assumptions can simply ignore, or warp the calling of tri- or tetra-alleleic positions.</p>
<p>Thus my script is responsible for generating a SAM file, describing the alignment of the generated reads against a reference that does not exist, and a VCF, which simply enumerates all genomic positions on the triviome as a potential variant.<br />
It&#8217;s important to note here that for <code>Gretel</code> (and recovery in general) the actual content of the reference sequence is irrelevant: the job of the reference (or <em>pseudo-reference</em>, as I have taken to call it for metagenomes) is to provide a shared co-ordinate system for the sequenced reads via alignment. In this case, the co-ordinates of the reads are known (we generated them!) and so the process of alignment is redundant and the reference need not exist at all.</p>
<p>Indeed, this framework has proved valuable. A harness script now rapidly and repeatedly generates hundreds of triviomes. Each test creates a number of haplotypes, with some number of SNPs. The harness can also specify an error rate, and how to sample the reads from each generated haplotype (uniformly, exponentially etc.). The resulting read alignment and variant list is thrown at <code>Hansel</code> and <code>Gretel</code> as part of the harness. Haplotypes recovered and output by <code>Gretel</code> can then be compared to the randomly generated input sequences for accuracy by merely constructing a matrix of Hamming distances (<em>i.e.</em> how many bases do not match for each output-input pair?). This is simple, fast and pretty effective, even if I do say so myself.</p>
<p>Despite its simplicity, this framework forms a basis for testing the packages during development, as well as giving us a platform on which to investigate the influence on recovery rate that parameters such as read length, number of haplotypes, number of SNPs, error rate, alignment rate have. Neat!</p>
<h3>Synthetic Metagenomes</h3>
<p>Of course, this triviome stuff is all well and good, but we need to prove that recovery is also possible on real genomic data. So we still left executing the rather convoluted looking workflow that I left you with at towards the end of my last report?</p>
<p>On the surface, that appears to be the case. Indeed, we must still simulate reads from some set of related genes, align those reads to some pseudo-reference, call for SNPs on that alignment and use the alignment and those called SNPs as the input to <code>Gretel</code> to recover the original genes that we generated reads from in the first place. Of course, we must also compare the haplotypes returned by <code>Gretel</code> to the known genes to evaluate the results.</p>
<p>But actually, the difficulty in the existing workflow is in the evaluation. Currently we use an alignment step to determine where each input gene maps to the selected pseudo-reference. This alignment is independent from the alignment of generated reads to the pseudo-reference. The hit table that describes how each input gene maps to the master is actually parsed as part of the evaluation step. To compare the input sequences against the recovered haplotypes, we need to know which parts of the input sequence actually overlap the recovered sequences (which share the same co-ordinates as the pseudo-reference), and where the start and end of that overlapping region exists on that particular input. The hit table effectively allows us to convert the co-ordinates of the recovered haplotypes, to those of the input gene. We can then compare bases by position and count how many match (and how many don&#8217;t).</p>
<p>Unsurprisingly, this got messy quite quickly, and the situation was exacerbated by subtle disagreements between the alignments of genes to the reference with <code>BLAST</code> and reads to the reference with <code>bowtie2</code>. This caused quite a bit of pain and ultimately ended with me manually inspecting sequences and writing my own <code>BLAST</code>-style hit tables to correct the evaluation process.</p>
<p>One afternoon whilst staring at <code>Tablet</code> and pondering my life choices, I wondered why I was even comparing the input and output sequences in this way. We&#8217;re effectively performing a really poor local alignment search between the input and output sequences. Using an aligner such as <code>BLAST</code> to compare the recovered haplotypes to the input sequences seems to be a rather intuitive idea. So why don&#8217;t we just do an actual local alignment?</p>
<p>Without a good answer, I tore my haplotype evaluation logic out of <code>Gretel</code> and put it in a metaphorical skip. Now we&#8217;ve dramatically simplified the process for generating and evaluating data sets. Hooray.</p>
<p>A bunch of small data sets now exist at <a href="https://github.com/SamStudio8/gretel-test">github.com/SamStudio8/gretel-test</a> and a framework of questionable bash scripts make the creation of new data sets rather trivial. Wonderful.</p>
<h2>Results</h2>
<p>So does this all actually work? Without wanting to step on the toes of my next blog post, it seems to, yes.</p>
<p>Accuracy across the triviome harness in general, is very good. Trivial haplotypes with up to 250 SNPs can be recovered <strong>in full</strong>, even in haplomes consisting of reads from 10 distinct, randomly generated haplotypes. Unsurprisingly, we&#8217;ve confirmed that increasing the number of haplotypes, and the number of SNPs on those haplotypes makes recovery more difficult.</p>
<p>To investigate recovery from metahaplomes of real genes, I&#8217;ve followed my previous protocol: select some gene (I&#8217;ve chosen <em>DHFR</em> and <em>AIMP1</em>), use <code>BLAST</code> to locate five arbitrary but similar genes from the same family, break them into reads and feed the alignment and called SNPs to <code>Gretel</code> with the goal of recovering the original genes. For both the DHFR and AIMP1 data sets, it is possible to get recovery rates of 95-100% for genes that look similar to the psuedo-reference and 80+% for those that are more dissimilar.</p>
<p>The relationship between pseudo-reference similarity and haplotype recovery rates might appear discouraging at first, but after digging around for the reasoning behind this result, it turns out not to be <code>Gretel</code>&#8216;s fault. Reads generated from sequences that have less identity to the pseudo-reference are less likely to align to that pseudo-reference, and are more likely to be discarded. <code>bowtie2</code> denies <code>Gretel</code> access to critical evidence required to accurately recover these less similar haplotypes.</p>
<p>This finding echoes an overarching theme I have encountered with current genomic tools and pipelines: not only are <strong>our current protocols and software not suitable for the analysis of metagenomes</strong>, but their underlying assumptions of diploidy are actually detrimental to the analyses we are conducting.</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">Making a start on the new paper <a href="https://t.co/dlDKrGKUy2">pic.twitter.com/dlDKrGKUy2</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/763341839474757632">August 10, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h2>Introducing our Preprint</h2>
<p>My work on everything so far has culminated in the production of a preprint: <a href="http://biorxiv.org/content/early/2016/08/02/067215">Advances in the recovery of haplotypes from the metagenome</a>. It&#8217;s quite humbling to see the sum total of around 18 months of my life summed into a document. Flicking through it a few months after it went online, I still get a warm fuzzy feeling: it looks like real science! I provide a proper definition of the <strong>metahaplome</strong> and introduce both the underlying graph theory for <code>Hansel</code> and the probability equations for <code>Gretel</code>. It also goes into a whole heap of results obtained so far, some historical background into the problem and where we are now as a field, and an insight into how this approach is different from other methodologies.</p>
<p><strong>Our work is an advance in computational methods for extracting exciting exploitable enzymes from metagenomes.</strong></p>
<hr />
<h1>Conclusion</h1>
<p>Things appear to be going well. Next steps are to get in the lab and try <code>Gretel</code> out for real. We are still hunting around for some DNA in a freezer that has a corresponding set of good quality sequenced reads.</p>
<p>In terms of development for <code>Hansel</code> and <code>Gretel</code>, there is still plenty of room for future work (helpfully outlined by my preprint) but in particular I still need to come up with a good way to incorporate quality scores and paired end information. I expect both attributes will improve performance more than the pretty awesome results we are getting already.</p>
<p>For more heavy refactoring, I&#8217;d like to also look at replacing <code>Gretel</code>&#8216;s inherent greedy bias (the best choice is always the edge with the highest probability) with something a little more clever. Additionally, the handling of indels is more than certainly going to become the next thorn in my side.</p>
<p>I&#8217;d also like to come up with some way of visualising results (I suspect it will involve <code>Circos</code>), because everybody can get behind a pretty graph.</p>
<hr />
<h1>In other news</h1>
<h4><strong>I went in the lab for the first time&#8230;</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">AUTHORISED TO WORK IN A CLASS 2 GM LAB, GET READY FOR SCIENCE</p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/743049192034172928">June 15, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>&#8230;it was not a success</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">LOOK AT THIS SHIT <a href="https://t.co/jkdIVxeZVz">pic.twitter.com/jkdIVxeZVz</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/745645929001254912">June 22, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>Biologists on Reddit actually liked my sassy PCR protocol</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">omg the real biologists are talking to me like a NOT PRETEND BIOLOGIST<a href="https://t.co/QDpneYoDBY">https://t.co/QDpneYoDBY</a> <a href="https://t.co/lzAb1TdLpX">pic.twitter.com/lzAb1TdLpX</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/745395241017090049">June 21, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>Our sys admin left</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">Trusty sysadmin couldn&#8217;t give root, but a nice dev machine. Here&#8217;s the MOTD. Thanks and good luck, <a href="https://twitter.com/MartinjVickers">@MartinjVickers</a> <a href="https://t.co/1YMXZxfcVL">pic.twitter.com/1YMXZxfcVL</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/733636009544536065">May 20, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>This happened.</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">SO WE AREN&#8217;T IN EUROPE AND NOW TGAC IS CALLED THE EARLHAM INSTITUTE I DON&#8217;T UNDERSTAND ANYTHING ANYMORE</p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/747341129545224193">June 27, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>Someone is actually using Goldilocks</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">SOMEONE IS USING MY SOFTWARE :&#8217;) <a href="https://t.co/tBo02HfyL0">https://t.co/tBo02HfyL0</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/751062183694434304">July 7, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>I am accidentally in charge of the department 3D printer</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">YESSSSSS IT&#8217;S HERE <a href="https://t.co/NjO0afZcgA">pic.twitter.com/NjO0afZcgA</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/752858081483165696">July 12, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>Writing a best man&#8217;s speech is definitely harder than an academic paper</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">I WAS AT A WEDDING <a href="https://t.co/t36bW1XIE6">pic.twitter.com/t36bW1XIE6</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/755160948105744384">July 18, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>I am a proper biologist now</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">I FEAR I AM NOT COMPETITIVE ENOUGH FOR MY ENVIRONMENT, I AM OVERRUN, SEND HELP HUMANS <a href="https://t.co/IQF078qxud">pic.twitter.com/IQF078qxud</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/758376448340160512">July 27, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>Someone trusts me to play with lasers</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="pl">PEW PEW <a href="https://t.co/JEXsoffBFm">pic.twitter.com/JEXsoffBFm</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/760793802491719680">August 3, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>I acquired some bees</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">AWAITING DUMPER TRUCK OF BEES <a href="https://twitter.com/hashtag/BEES?src=hash">#BEES</a> <a href="https://twitter.com/DarkestKale">@DarkestKale</a> <a href="https://t.co/s5xEr1aexT">pic.twitter.com/s5xEr1aexT</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/761848216719679488">August 6, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>I went up some mountains with a nice man from the internet</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p dir="ltr" lang="en">Mountain 1 complete <a href="https://t.co/9CvgSsmCwg">pic.twitter.com/9CvgSsmCwg</a></p>
<p>— Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/764432713440956417">August 13, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>I went to a conference&#8230;</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">At the <a href="https://twitter.com/eccb2016">@eccb2016</a> welcome party. The drinks don&#39;t appear to be limited, so I&#39;m on my fifth wine. Cheers :3 <a href="https://t.co/QBTjwmIMYY">pic.twitter.com/QBTjwmIMYY</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/772503338537680896">September 4, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>&#8230;I presented some work!</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">Visit DA125 to hear how to pull exciting enzymes out of metagenomes before I have too much free wine <a href="https://twitter.com/hashtag/eccb2016?src=hash">#eccb2016</a> <a href="https://t.co/7o5ueuXLnS">pic.twitter.com/7o5ueuXLnS</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/772827884067123200">September 5, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>I accidentally went on holiday</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">Me looking at email on holiday <a href="https://t.co/5FLWZamsnZ">pic.twitter.com/5FLWZamsnZ</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/775223679106027520">September 12, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<h4><strong>&#8230;and now I am back.</strong></h4>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">Urgh. <a href="https://t.co/vzRZmq7xuw">pic.twitter.com/vzRZmq7xuw</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/781057080815919104">September 28, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<hr />
<h1>tl;dr</h1>
<ul>
<li>Yes hello, I am still here and appear to be worse at both blog and unit testing</li>
<li>My work now has a name: a data structure called <a href="https://github.com/SamStudio8/hansel/"><code>Hansel</code></a> and an algorithm called <a href="https://github.com/SamStudio8/gretel/"><code>Gretel</code></a></li>
<li>Reweighting appears to work in a robust fashion now that it is implemented correctly</li>
<li>Smoothing the conditional probabilities of observations should be looked at again</li>
<li><em>Triviomes</em> provide a framework for investigating the effects of various aspects of metagenomes on the success of haplotype recovery</li>
<li>Generating synthetic metahaplomes for testing is significantly less of a pain in the ass now I have simplified the process for evaluating them</li>
<li><strong>Current protocols and software are not suitable for the analysis of metagenomes</strong> (apart from mine)</li>
<li><a href="http://biorxiv.org/content/early/2016/08/02/067215">I have a preprint</a>!</li>
<li>Indels are almost certainly going to be the next pain in my ass</li>
<li>I shunned my responsibilities as a PhD student for a month and travelled Eastern Europe enjoying the sun and telling strangers about how awful bioinformatics is</li>
<li>I am now back telling North Western Europeans how awful bioinformatics is</li>
</ul>
<div class="footnotes">
<hr />
<ol>
<li id="fn-1201-1">
Mhaplotyper? Oh yes, the M is for metagenomic, and it is also silent.&#160;<a href="#fnref-1201-1" rev="footnote">&#8617;</a>
</li>
<li id="fn-1201-2">
Having written this paragraph, I wonder what the real impact of this intuition actually is. I&#8217;ve now <a href="https://github.com/SamStudio8/gretel/issues/3">opened an issue on my own repo</a> to track my investigation. In the case where more than <code>L</code> pairs of evidence exist, is there quantifiable loss in accuracy by only considering <code>L</code> pairs of evidence? In the case where fewer than <code>L</code> pairs of evidence exist, does the smoothing have a non-negligible affect on performance?&#160;<a href="#fnref-1201-2" rev="footnote">&#8617;</a>
</li>
<li id="fn-1201-3">
Of course, the concept of alignment in a <em>triviome</em> is somewhat undefined as we have no &#8220;reference&#8221; sequence anyway. Although one could select one of the random haplotypes as a pseudo-reference against which to align all tiny reads, considering the very short read lengths and the low sequence identity between the randomly generated haplotype sequences, it is highly likely that the majority of reads will fail to align correctly (if at all) to such a reference.&#160;<a href="#fnref-1201-3" rev="footnote">&#8617;</a>
</li>
</ol>
</div>
</div>
                
                        <div class="tag-well"><span><span class="fa fa-tags"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/tag/metahaplome/" rel="tag">metahaplome</a>,&nbsp; <a href="https://samnicholls.net/tag/paper/" rel="tag">paper</a>,&nbsp; <a href="https://samnicholls.net/tag/preprint/" rel="tag">preprint</a>,&nbsp; <a href="https://samnicholls.net/tag/sitrep/" rel="tag">sitrep</a>,&nbsp; <a href="https://samnicholls.net/tag/status-report/" rel="tag">status report</a></span></div>
                
                <nav class="further-reading">
	<div class="previous">
		<span>Previous Post</span>
		<a href="https://samnicholls.net/2016/06/26/edta-buffer-protocol/">A pretend biologist&#8217;s guide to making an EDTA chelating (&#8220;buffer&#8221;) solution</a>
	</div>
	<div class="next">
		<span>Next Post</span>
		<a href="https://samnicholls.net/2016/11/03/talks-and-metaphors/">Interdisciplinary talks and the metaphor-ome: Harder than metagenomics itself?</a>
	</div>
</nav>            </article>
                                <div class="card">

<div id="disqus_thread"></div>
                </div>
    
</section>

        <!-- Sidebar -->
        <aside class="col-md-4 sidebar-container">
		            <div id="widget-area" class="widget-area" role="complementary">
                <div id="custom_html-3" class="widget_text card widget widget_custom_html"><div class="textwidget custom-html-widget"><img src="https://pbs.twimg.com/profile_images/1044255545610436609/wt33ZbtU_400x400.jpg" height=120 style="margin-top: 20px; margin-right:10px; float: left;" alt="Sam Nicholls" />
<div style="padding-top:15px; line-height: normal"><b style="font-size: 1.2em">Sam Nicholls</b><br><span style="font-size: 0.9em;">Haplotype Wrangler at University of Birmingham.<br>I tell computers to do things, then I write about how it all went wrong.</span></div></div></div><div id="search-2" class="card widget widget_search"><form style="margin-top:15px" action="https://samnicholls.net/" id="searchform" method="get" role="form">
    <div class="form-group">
        <input type="text" class="form-control" name="s" id="s" placeholder="e.g. Search...">
    </div>
</form>
</div><div id="twitter_timeline-5" class="card widget widget_twitter_timeline"><a class="twitter-timeline" data-height="400" data-theme="light" data-link-color="#f96e5b" data-border-color="#e8e8e8" data-lang="EN" data-partner="jetpack" data-chrome="noheader nofooter noborders" href="https://twitter.com/samstudio8">My Tweets</a></div><div id="categories-5" class="card widget widget_categories"><h2>Categories</h2>
			<ul>
					<li class="cat-item cat-item-2"><a href="https://samnicholls.net/category/bioinf/">Bioinformatics</a>
<ul class='children'>
	<li class="cat-item cat-item-8"><a href="https://samnicholls.net/category/bioinf/au-phd/">AU-PhD</a>
</li>
	<li class="cat-item cat-item-3"><a href="https://samnicholls.net/category/bioinf/sanger-qc/" title="Sanger QC Project">Sanger-QC</a>
</li>
	<li class="cat-item cat-item-61"><a href="https://samnicholls.net/category/bioinf/tools/">Tools</a>
	<ul class='children'>
	<li class="cat-item cat-item-185"><a href="https://samnicholls.net/category/bioinf/tools/chitin/">chitin</a>
</li>
	</ul>
</li>
</ul>
</li>
	<li class="cat-item cat-item-80"><a href="https://samnicholls.net/category/devops/">Devops</a>
</li>
	<li class="cat-item cat-item-114"><a href="https://samnicholls.net/category/event/">Event</a>
</li>
	<li class="cat-item cat-item-21"><a href="https://samnicholls.net/category/gadgets/">Gadgets</a>
</li>
	<li class="cat-item cat-item-176"><a href="https://samnicholls.net/category/hardware/">Hardware</a>
<ul class='children'>
	<li class="cat-item cat-item-223"><a href="https://samnicholls.net/category/hardware/beehive-monitor/">Beehive Monitor</a>
</li>
	<li class="cat-item cat-item-178"><a href="https://samnicholls.net/category/hardware/esp8266/">ESP8266</a>
</li>
	<li class="cat-item cat-item-236"><a href="https://samnicholls.net/category/hardware/lego-sequencer/">Lego Sequencer</a>
</li>
	<li class="cat-item cat-item-177"><a href="https://samnicholls.net/category/hardware/lightstick/">Lightstick</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-151"><a href="https://samnicholls.net/category/lab/">Lab</a>
<ul class='children'>
	<li class="cat-item cat-item-150"><a href="https://samnicholls.net/category/lab/protocol-guides/">Protocol Guides</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-5"><a href="https://samnicholls.net/category/meta/">Meta</a>
</li>
	<li class="cat-item cat-item-62"><a href="https://samnicholls.net/category/mysteries/">Mysteries</a>
</li>
	<li class="cat-item cat-item-22"><a href="https://samnicholls.net/category/personal/">Personal</a>
</li>
	<li class="cat-item cat-item-4"><a href="https://samnicholls.net/category/photo/">Photography</a>
</li>
	<li class="cat-item cat-item-143"><a href="https://samnicholls.net/category/programming/">Programming</a>
<ul class='children'>
	<li class="cat-item cat-item-145"><a href="https://samnicholls.net/category/programming/documentation/">Documentation</a>
</li>
	<li class="cat-item cat-item-144"><a href="https://samnicholls.net/category/programming/python/">Python</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-48"><a href="https://samnicholls.net/category/status-report/">Status Report</a>
</li>
	<li class="cat-item cat-item-60"><a href="https://samnicholls.net/category/sysadm/">System Administration</a>
</li>
	<li class="cat-item cat-item-1"><a href="https://samnicholls.net/category/uncategorised/">Uncategorised</a>
</li>
			</ul>

			</div>            </div>
		        </aside>
    </div>

</div> <!-- /container -->

<footer id="main-footer" class="bg-primary">
	<div id="footer-widgets" class="container">
		    </div>
</footer>



	<div style="display:none">
	</div>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/comment-reply.min.js?ver=5.7.5' id='comment-reply-js'></script>
<script type='text/javascript' id='disqus_count-js-extra'>
/* <![CDATA[ */
var countVars = {"disqusShortname":"samposium-blog"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_count.js?ver=3.0.17' id='disqus_count-js'></script>
<script type='text/javascript' id='disqus_embed-js-extra'>
/* <![CDATA[ */
var embedVars = {"disqusConfig":{"integration":"wordpress 3.0.17"},"disqusIdentifier":"1201 https:\/\/samnicholls.net\/?p=1201","disqusShortname":"samposium-blog","disqusTitle":"Status Report: Jul Aug September 2016","disqusUrl":"https:\/\/samnicholls.net\/2016\/10\/24\/status-sep16\/","postId":"1201"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_embed.js?ver=3.0.17' id='disqus_embed-js'></script>
<script type='text/javascript' src='https://s0.wp.com/wp-content/js/devicepx-jetpack.js?ver=202335' id='devicepx-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/jetpack/_inc/twitter-timeline.js?ver=4.0.0' id='jetpack-twitter-timeline-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/wp-embed.min.js?ver=5.7.5' id='wp-embed-js'></script>
<script type='text/javascript' src='https://stats.wp.com/e-202335.js' async defer></script>
<script type='text/javascript'>
	_stq = window._stq || [];
	_stq.push([ 'view', {v:'ext',j:'1:4.9',blog:'101350222',post:'1201',tz:'1',srv:'samnicholls.net'} ]);
	_stq.push([ 'clickTrackerInit', '101350222', '1201' ]);
</script>

<script type="text/javascript">
(function() {
  if (navigator.userAgent.match(/IEMobile\/10\.0/)) {
    var msViewportStyle = document.createElement("style");
    msViewportStyle.appendChild(
      document.createTextNode("@-ms-viewport{width:auto!important}")
    );
    document.getElementsByTagName("head")[0].appendChild(msViewportStyle);
  }
})();

var collapsed = false;
jQuery( document ).ready(function($) {
    $('.navbar-toggle').click(function(){
        console.log(collapsed);
        if(collapsed) {
            $('.navbar-header > button').html('<i class="fa fa-bars"></i>');
        } else {
            $('.navbar-header > button').html('<i class="fa fa-times"></i>');
        }
        $('.navbar-collapse').toggle(200);
        collapsed = !collapsed;
    });
});
</script>
	
</body>

</html>
