<!DOCTYPE html>
<html lang="en-GB">

<head>
	<meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="pingback" href="https://samnicholls.net/xmlrpc.php">

	
	<title>Status Report: May 2016 (Metahaplomes: The graph that isn&#8217;t) &#8211; Samposium</title>
<meta name='robots' content='max-image-preview:large' />
<link rel='dns-prefetch' href='//s0.wp.com' />
<link rel='dns-prefetch' href='//fonts.googleapis.com' />
<link rel='dns-prefetch' href='//s.w.org' />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Feed" href="https://samnicholls.net/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Comments Feed" href="https://samnicholls.net/comments/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Status Report: May 2016 (Metahaplomes: The graph that isn&#8217;t) Comments Feed" href="https://samnicholls.net/2016/06/12/status-may16/feed/" />
		<script type="text/javascript">
			window._wpemojiSettings = {"baseUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/72x72\/","ext":".png","svgUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/svg\/","svgExt":".svg","source":{"concatemoji":"https:\/\/samnicholls.net\/wp-includes\/js\/wp-emoji-release.min.js?ver=5.7.5"}};
			!function(e,a,t){var n,r,o,i=a.createElement("canvas"),p=i.getContext&&i.getContext("2d");function s(e,t){var a=String.fromCharCode;p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,e),0,0);e=i.toDataURL();return p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,t),0,0),e===i.toDataURL()}function c(e){var t=a.createElement("script");t.src=e,t.defer=t.type="text/javascript",a.getElementsByTagName("head")[0].appendChild(t)}for(o=Array("flag","emoji"),t.supports={everything:!0,everythingExceptFlag:!0},r=0;r<o.length;r++)t.supports[o[r]]=function(e){if(!p||!p.fillText)return!1;switch(p.textBaseline="top",p.font="600 32px Arial",e){case"flag":return s([127987,65039,8205,9895,65039],[127987,65039,8203,9895,65039])?!1:!s([55356,56826,55356,56819],[55356,56826,8203,55356,56819])&&!s([55356,57332,56128,56423,56128,56418,56128,56421,56128,56430,56128,56423,56128,56447],[55356,57332,8203,56128,56423,8203,56128,56418,8203,56128,56421,8203,56128,56430,8203,56128,56423,8203,56128,56447]);case"emoji":return!s([55357,56424,8205,55356,57212],[55357,56424,8203,55356,57212])}return!1}(o[r]),t.supports.everything=t.supports.everything&&t.supports[o[r]],"flag"!==o[r]&&(t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&t.supports[o[r]]);t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&!t.supports.flag,t.DOMReady=!1,t.readyCallback=function(){t.DOMReady=!0},t.supports.everything||(n=function(){t.readyCallback()},a.addEventListener?(a.addEventListener("DOMContentLoaded",n,!1),e.addEventListener("load",n,!1)):(e.attachEvent("onload",n),a.attachEvent("onreadystatechange",function(){"complete"===a.readyState&&t.readyCallback()})),(n=t.source||{}).concatemoji?c(n.concatemoji):n.wpemoji&&n.twemoji&&(c(n.twemoji),c(n.wpemoji)))}(window,document,window._wpemojiSettings);
		</script>
		<style type="text/css">
img.wp-smiley,
img.emoji {
	display: inline !important;
	border: none !important;
	box-shadow: none !important;
	height: 1em !important;
	width: 1em !important;
	margin: 0 .07em !important;
	vertical-align: -0.1em !important;
	background: none !important;
	padding: 0 !important;
}
</style>
	<link rel='stylesheet' id='crayon-css'  href='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/css/min/crayon.min.css?ver=_2.7.2_beta' type='text/css' media='all' />
<link rel='stylesheet' id='wp-block-library-css'  href='https://samnicholls.net/wp-includes/css/dist/block-library/style.min.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='child-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='lato-css'  href='//fonts.googleapis.com/css?family=Lato%3A400%2C700%2C400italic&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='slabo-css'  href='//fonts.googleapis.com/css?family=Slabo+27px&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='fa-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/font-awesome.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='rd-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-custom-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/custom.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='jetpack_css-css'  href='https://samnicholls.net/wp-content/plugins/jetpack/css/jetpack.css?ver=4.9' type='text/css' media='all' />
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery.min.js?ver=3.5.1' id='jquery-core-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery-migrate.min.js?ver=3.3.2' id='jquery-migrate-js'></script>
<script type='text/javascript' id='crayon_js-js-extra'>
/* <![CDATA[ */
var CrayonSyntaxSettings = {"version":"_2.7.2_beta","is_admin":"0","ajaxurl":"https:\/\/samnicholls.net\/wp-admin\/admin-ajax.php","prefix":"crayon-","setting":"crayon-setting","selected":"crayon-setting-selected","changed":"crayon-setting-changed","special":"crayon-setting-special","orig_value":"data-orig-value","debug":""};
var CrayonSyntaxStrings = {"copy":"Press %s to Copy, %s to Paste","minimize":"Click To Expand Code"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/js/min/crayon.min.js?ver=_2.7.2_beta' id='crayon_js-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/fd-footnotes/fdfootnotes.js?ver=1.34' id='fdfootnote_script-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/bootstrap.js?ver=5.7.5' id='bootstrap-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/custom.js?ver=5.7.5' id='js-custom-js'></script>
<link rel="https://api.w.org/" href="https://samnicholls.net/wp-json/" /><link rel="alternate" type="application/json" href="https://samnicholls.net/wp-json/wp/v2/posts/699" /><link rel="EditURI" type="application/rsd+xml" title="RSD" href="https://samnicholls.net/xmlrpc.php?rsd" />
<link rel="wlwmanifest" type="application/wlwmanifest+xml" href="https://samnicholls.net/wp-includes/wlwmanifest.xml" /> 
<meta name="generator" content="WordPress 5.7.5" />
<link rel="canonical" href="https://samnicholls.net/2016/06/12/status-may16/" />
<link rel='shortlink' href='https://wp.me/p6RfP0-bh' />
<link rel="alternate" type="application/json+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F06%2F12%2Fstatus-may16%2F" />
<link rel="alternate" type="text/xml+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F06%2F12%2Fstatus-may16%2F&#038;format=xml" />
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-62206786-1', 'auto');
  ga('send', 'pageview');

</script>


<link rel='dns-prefetch' href='//v0.wordpress.com'>
<style type='text/css'>img#wpstats{display:none}</style>
<!-- Jetpack Open Graph Tags -->
<meta property="og:type" content="article" />
<meta property="og:title" content="Status Report: May 2016 (Metahaplomes: The graph that isn&#8217;t)" />
<meta property="og:url" content="https://samnicholls.net/2016/06/12/status-may16/" />
<meta property="og:description" content="Another long overdue insight to the things I am supposed to be doing so I can call myself a doctor someday." />
<meta property="article:published_time" content="2016-06-12T20:03:24+00:00" />
<meta property="article:modified_time" content="2016-06-12T20:21:53+00:00" />
<meta property="og:site_name" content="Samposium" />
<meta property="og:image" content="https://s0.wp.com/i/blank.jpg" />
<meta property="og:locale" content="en_GB" />
<meta name="twitter:card" content="summary" />
</head>

<body class="post-template-default single single-post postid-699 single-format-standard">


        
<header id="main-header">
    <nav class="navbar-default navbar-static-top navbar" role="navigation">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#rd-collapse">
                    <span class="fa fa-bars"></span>
                    <span class="sr-only screen-reader-text">Toggle navigation</span>
                </button>
                <a class="navbar-brand" href=" https://samnicholls.net/">Samposium</a>            </div>
            
            <div id="rd-collapse" class="collapse navbar-collapse">
            
            <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body class="post-template-default single single-post postid-699 single-format-standard">

<div id="main-menu" class="navbar-nav nav"><ul>
<li class="page_item page-item-8"><a href="https://samnicholls.net/about/">About Me</a></li>
<li class="page_item page-item-11"><a href="https://samnicholls.net/talks/">Talks</a></li>
<li class="page_item page-item-15"><a href="https://samnicholls.net/things-i-have-learned/">Things I Have Learned</a></li>
</ul></div></body></html>
        </div>

        </div>
    </nav>
</header>

<div class="container">

    <div class="row"><!-- Content -->
<section class="col-md-8 content-container">
                          <article class="card post hentry post-699 type-post status-publish format-standard category-au-phd category-status-report tag-graph tag-metagenome tag-metahaplome tag-sitrep tag-status-report" id="post-699">
                <header class="entry-header">
                        <h1 class="entry-title">Status Report: May 2016 (Metahaplomes: The graph that isn&#8217;t)</h1>                </header><!-- .entry-header -->

                <div class="meta-well">
                    <span style="margin-right:15px;" class=""><span class="fa fa-user"></span>&nbsp;&nbsp;&nbsp;<span class="author vcard"><a class="url fn n" href="https://samnicholls.net/about/">Sam Nicholls</a></span></span>

                    <span style="margin-right:15px;" class="entry-date published updated post-date"><span class="fa fa-clock-o"></span>&nbsp;&nbsp; <time class="entry-date published" datetime="2016-06-12T21:03:24+01:00">12th June 2016</time><time class="updated" datetime="2016-06-12T21:21:53+01:00">12th June 2016</time></span>                    <span style="margin-right:15px;" class=""><span class="fa fa-comments-o"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/2016/06/12/status-may16/#respond"><span class="dsq-postid" data-dsqidentifier="699 https://samnicholls.net/?p=699">No Comments yet</span></a></span>
                    <span class=""><span class="fa fa-folder-open"></span>&nbsp;&nbsp;&nbsp;<a href="https://samnicholls.net/category/bioinf/au-phd/" rel="category tag">AU-PhD</a>, <a href="https://samnicholls.net/category/status-report/" rel="category tag">Status Report</a></span>
                </div>

                <div class="entry-content"><p>It would seem that a sufficient amount of time has passed since my <a href="https://samnicholls.net/2016/03/01/status-feb16/">previous report</a> to discuss how everything has broken in the meantime. You would have left off with a version of me who had not long solidified the concept of the <strong>metahaplome</strong>: a graph-inspired representation of the variation observed across aligned reads from a sequenced metagenome. Where am I now?</p>
<h2>Metahaplomes</h2>
<h3>The graph that isn&#8217;t</h3>
<p>At the end of my first year, I returned from my <a href="https://samnicholls.net/2015/08/04/belgian/">reconnaissance mission to extract data mining knowledge from a leading Belgian university</a> with a prototype for a data structure that was fit to house a metahaplome; a probabilistically weighted <em>graph</em> that can be traversed to extract likely sequences of variants on some gene of interest. I say <em>graph</em>, because the structure and API does not look or particularly act like a traditional graph at all. Indeed, the current representation is a four dimensional matrix that stores the number of observations of a symbol (SNP) <em>A</em> at position <em>i</em>, co-occurring with a symbol <em>B</em> at position <em>j</em>.</p>
<p>This has proved problematic as I&#8217;ve had difficulty in explaining the significance of this to people who dare to ask what my project is about. &#8220;What do you mean it&#8217;s <strong>not</strong> a graph? There&#8217;s a picture of a graph on your poster right there!?&#8221;. Yes, the matrix can be exploited to build a simple graph representation, but not without some information loss. As a valid gene must select a variant at each site, one cannot draw a graph that contains edges from sites of polymorphisms that are not adjacent (as a path that traverses such an edge would skip a variant site<sup id="fnref-699-1"><a href="#fn-699-1" rel="footnote">1</a></sup>). We therefore lose the ability to encode any information regarding co-occurrence of non-adjacent variants (<code>abs(i - j) != 1</code>) if we depict the problem with a simple graph alone.</p>
<p>To circumvent this, edges are not weighted upfront. Instead, to take advantage of the evidence available, the graph is dynamically weighted <strong>during</strong> traversal (the movement to the next node is variable, and depends on the nodes that have been visited already) using the elements stored in the matrix.</p>
<p>Thus we have a data structure capable of being utilised <em>like</em> a graph, with some caveats: it is not possible to enumerate all possibilities or assign weights to all edges upfront before traversal (or for that matter, a random edge), and a fog of war exists during any traversal (<em>i.e.</em> it is not possible to predict where a path may end without exploring). Essentially we have no idea what the graph looks like, until we explore it. Despite this, my solution fuses the advantage of a graph&#8217;s simple representation, with the advantage of an adjacency matrix that permits storage of all pertinent information. Finally, I&#8217;ve been able to describe the structure and algorithm verbally and mathematically.</p>
<h3>Reweighting</h3>
<p>Of course, having this traversable structure that takes all the evidence seen across the reads into account is great, but we need a reliable method for rescuing more than just one possible gene variant from the target metahaplome. My initial attempts at this involved invoking stochastic jitter during traversal to quite poor effect. It was not until some time after I&#8217;d got back from putting mayonnaise on everything that I considered altering the observation matrix that backs the graph itself to achieve this.</p>
<p>My previous report described the current methodology: given a complete path, check the marginal probability for each variant at each position of the path (<em>i.e.</em> the probability one would select the same nucleotide if you were to look at variant site in isolation) and determine the smallest marginal. Then iterate over the path, down-weighting the element of the observation matrix that stores the number of occurrences of the <code>i</code>&#8216;th nucleotide and the <code>i+1</code>&#8216;th selected nucleotide, by multiplying the existing value by the lowest marginal (which will be greater than 0, but smaller than 1) and subtracting that value from the current count.</p>
<p>Initial testing yielded more accurate results with this method than anything I had tried previously, where accuracy is quantified by this not happening:</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">oh no dont tell my supervisors <a href="https://t.co/7j6C2oS8BD">pic.twitter.com/7j6C2oS8BD</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/680543717493714944">December 26, 2015</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p>The algorithm is evaluated with a data set of several known genes from which a metagenome is simulated. The coloured lines on the chart above refer to each known input gene. The <em>y</em> axis represents the percentage of variants that are &#8220;recovered&#8221; from the metagenome, the <em>x</em> axis is the iteration (or path) number. In this example, a questionable strategy caused poor performance (other than the 100% recovery of the blue gene), and a bug in handling elements that are reweighted below 1 allowed the algorithm to enter a periodic state.</p>
<p>After implementing the latest strategy, performance compared to the above increased significantly (at least on the limited data sets I have spent the time curating), but I was still not entirely satisfied. Recognising this was going to take much more time and thought, I procrastinated by writing up the technical aspects of my work in excruciating mathematical detail in preparation for my next paper. To wrap my head around my own equations, I commandeered the large whiteboards in the undergraduate computing room and primed myself with coffee and <a href="https://play.spotify.com/album/5iLpRq7zbCArBm2Gi0pRIw">Galantis</a>. Unfortunately, after a hour or two of excited scribbling, this happened:</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">my day <a href="https://t.co/4Zhr5AHGY5">pic.twitter.com/4Zhr5AHGY5</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/727518138951176194">May 3, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p>I encountered an oversight. Bluntly:</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">.<a href="https://twitter.com/jhrtn">@jhrtn</a> it&#39;s a crap owl who is pissed off that the matrix it is guarding has been incorrectly re-weighted by a half baked algorithm i wrote</p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/727536307073892352">May 3, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p>Despite waxing lyrical about the importance of evidence arising from non-adjacent variant sites, I&#8217;d overlooked them in the reweighting process. Although frustrated with my own incompetence, this issue was uncovered at a somewhat opportune time as I was looking for a likely explanation for what felt like an upper bound on the performance of the algorithm. As evidence (observation counts) for adjacent pairwise variants was decreased through reweighting, non-adjacent evidence was becoming an increasingly important factor in the decision making process for path traversal, simply by virtue of the counts being larger (as they were left untouched). Thus paths were still being heavily coerced along particular routes and were not afforded the opportunity to explore more of the graph, yielding less accurate results (less recovered variants) for more divergent input genes.</p>
<p>As usual in these critical oversights, the fix was trivial (just ensure to apply the same rules for reweighting the adjacent pairs of variants to the non-adjacent ones too), and indeed, performance was bumped by around 5%p. Hooray.</p>
<h2>Evaluation</h2>
<h3>Generating test data (is still a pain in the arse)</h3>
<p>So here we are, I&#8217;m still somewhat stuck in a data rut. Generating data sets (that can be verified) is a somewhat convoluted procedure. Whilst, to run the algorithm all one needs is a BAM of aligned reads, and an associated VCF of called SNP sites; to empirically test output, we also need to know what the output genes should look like. Currently this requires a &#8220;master&#8221; FASTA (the origin gene), a FASTA of similar genes (the ones we actually want to recover) and a <code>blast</code> hit table that documents how those similar genes align to the master. The workflow for generating and testing a data set looks like this:</p>
<ul>
<li>Select an interesting, arbitrary <em>master</em> gene from a database (<code>master.fa</code>)</li>
<li><code>blast</code> for similar genes and select several hits with decreasing identity</li>
<li>Download FASTA (<code>genes.fa</code>) and associated <code>blast</code> hit table (<code>hits.txt</code>) for selected genes</li>
<li>Simulate reads by shredding <code>genes.fa</code> (<code>reads.fq</code>)</li>
<li>Align reads (<code>reads.fq</code>) with <code>bowtie</code> to pseudo-reference (<code>master.fa</code>) to create (<code>hoot.bam</code>)</li>
<li>Call for SNPs on (<code>hoot.bam</code>) to create a VCF (<code>hoot.vcf</code>)</li>
<li>Construct metahaplome and traverse paths with reads (<code>hoot.bam</code>) and SNPs (<code>hoot.vcf</code>)</li>
<li>Output potential genes (<code>out.fa</code>)</li>
<li>Evaluate each result in <code>out.fa</code> against each hit in <code>hits.txt</code>
<ul>
<li>Extract DNA between subject start and end for record from <code>genes.fa</code></li>
<li>Determine segment of output (from <code>out.fa</code>) overlapping current hit (from <code>genes.fa</code>)</li>
<li>Convert co-ordinates of SNP to current hit (<code>gene.fa</code>)</li>
<li>Confirm consistency between SNP on output, to </li>
<li>Return matrix of consistency for each output gene, to each input gene</li>
</ul>
</li>
</ul>
<h3>Discordant alignments between discordant aligners</h3>
<p>The issue at hand primarily arises from discordant alignment decisions between the two alignment processes that make up components of the pipeline; <code>blast</code> and <code>bowtie</code>. Although <code>blast</code> is used to select the initial genes (given some &#8220;master&#8221;), and its resulting hit table is also used to evaluate the approach at the end of the algorithm,  <code>bowtie</code> is used to align the reads (<code>reads.fq</code>) to that same master. Occasional disagreements between both algorithms are inevitable on real data, but I assumed that given the simplicity of the data (simulated reads of uniform length, no errors, reasonable identity) that they would behave the same. It may sound like an obvious problem source, but when several genes are reported as correctly extracted with high accuracy (identity) and one or two are not, you might forgive me for thinking that the algorithm just needed tweaking rather than an underlying problem stemming from the alignment steps! This led to much more tail chasing than I would care to admit to.</p>
<p>For one example: I investigated a poorly reconstructed gene by visualising the input BAM produced by <code>bowtie</code> with <a href="https://ics.hutton.ac.uk/tablet/">Tablet (a rather nifty BAM+VCF+FA viewer)</a>. It turned out that for input reads belonging to one of the input genes, <code>bowtie</code> had called an indel<sup id="fnref2:699-1"><a href="#fn-699-1" rel="footnote">1</a></sup>, causing a disagreement as to what the empirical base at each SNP following that indel should have been. That is, although all of the reads from that particular input gene were aligned by <code>bowtie</code> as having an indel (and thus shifting bases in those reads), and processed by my algorithm with that indel taken into account, at the point of evaluation, the <code>blast</code> hit table is the gold standard; what may have been the correct variant (indel withstanding) would be determined as incorrect by the alignment of the hit table.</p>
<p>I suppose the solution might be to switch to one aligner, but I&#8217;m aware that even the same aligner can make different decisions under differing conditions (read length).<br />
It&#8217;s important to note that currently the hit table is also used to define where to begin sharding reads for the simulated metagenome, which in turn causes trouble if <code>bowtie</code> disagrees with where an alignment begins and ends. I&#8217;ve had cases where <code>blast</code> aligns the first base of the subject (input gene) to the first base of the query (master) but on inspection with <code>Tablet</code>, it becomes evident that <code>bowtie</code> clips the first few bases when aligning opening reads to the master. This problem is a little more subtle, and in current practice causes little trouble. Although the effect would be a reduction in observed evidence for variants at SNPs that happen to occur within the first few bases of the gene, my test sets so far do not have a SNP site so close to the start of the master. This is obviously something to watch out for, though.</p>
<p>At this time I&#8217;ve just been manually altering the hit table to reconcile differences between the two aligners, which is gross.</p>
<h3>Bugs in the machine</h3>
<p>Of course, a status report from me is not complete without some paragraph where I hold my hands up in the air and say everything was broken due to my own incompetence. Indeed, a pair of off-by-one bugs in my evaluation algorithm also warped reported results. The first, a regression introduced after altering the parameters under which the evaluation function determines an overlap between the current output gene and current hit, led to a miscalculation when translating the base position on the master to the base position on the expected input under infrequent circumstances, causing the incorrect base to be compared to the expected output. This was also accidentally fixed when I refactored the code and saw a very small increase in performance.</p>
<p>The second, an off-by-one error in the reporting of a new metric: &#8220;deviations from reference&#8221;, caused the results to suddenly appear rather unimpressive. The metric measures the number of bases that are different from the pseudo-reference (<code>master.fa</code>) that were correctly recovered by my algorithm to match an original gene from <code>gene.fa</code>. Running my algorithm now yielded a results table describing impressive gene recovery scores (>89%) but those genes only appeared to differ from the reference by merely a few SNPs (&lt;10). How could we suck at recovering sequences that barely deviate from the master? Why does it take so many iterations? After getting off the floor and picking up the shattered pieces of my ego and PhD, I checked the VCF and confirmed there were over a hundred SNPs across all the genes. Curious, I inspected the genes manually with <code>Tablet</code> to see how they compared to the reference. Indeed, there were definitely more than the four reported for one particular case, so what was going on?</p>
<p>To finish quickly; path iteration numbers start from 0, but are reported to the user as <code>iter + 1</code>, because the 0&#8217;th iteration is not catchy. My mistake was using the <code>iter + 1</code> to also access the number of deviations from the reference detected in the current iteration &#8211; in a zero indexed structure. I was fetching the number of deviations successfully extracted by the path <em>after</em> the this one, which we would expect to be poor, as the structure would have been reweighted to prevent that path from appearing again. Nice work, me. This fix made things a little more interesting:</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">ONE HUNDRED PERCENT MATCH TO SECRET INPUT EXTRACTED, 118 SNPS, 84 DEVIATIONS FROM THE REFERENCE. ITSSS HAPPENINNGGG! <a href="https://t.co/KRdVaMT92U">pic.twitter.com/KRdVaMT92U</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/736779547069272064">May 29, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p>More testing of the testing, is evidently necessary.</p>
<h2>Conclusion</h2>
<p>So where does that leave us? Performance is up, primarily because the code that I wrote to evaluate performance (and reweight) is now less broken. Generating data sets is still a pain in the arse, but I have got the hang of the manual process involved so I can at least stop hiding from the work to be done. It might be worth investigating consolidating all of my alignment activities into one aligner to improve my credit score. Results are looking promising, this algorithm is now capable of extracting genes (almost or entirely whole) from simulated (albeit quite simple) metagenomes.</p>
<p>Next steps will be more testing, writing the method itself as a paper, and getting some proper biological evidence from the lab that this work can do what I tell people it can do.</p>
<h2>In other news</h2>
<ul>
<li>I&#8217;m actually a published scientist</li>
<li>We won some research funding to put me in a lab coat over summer and prove this works</li>
<li><a href="https://twitter.com/samstudio8/status/720658272890580992">Metahaplome poster won an award and represented the university at &#8220;Excellence with Impact&#8221;</a></li>
<li><a href="https://twitter.com/samstudio8/status/729396870678753286">I&#8217;ve managed to enjoy the sea</a></li>
<li><a href="https://twitter.com/samstudio8/status/729397562319482880">I&#8217;ve decided to be a beekeeper</a></li>
<li><a href="https://twitter.com/samstudio8/status/721328289592320001">I went to the Gregynog stats conference to enjoy an old house and talk about my work</a></li>
<li><a href="https://twitter.com/samstudio8/status/724984139141795844">I&#8217;ve done some lab work</a></li>
<li>I was invited to review my first paper (someone trusts what I say?)</li>
<li>I continue to succeed at pretending to be a biologist</li>
</ul>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">i like to pretend <a href="https://t.co/02WrtLf1fo">pic.twitter.com/02WrtLf1fo</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/732555637767458817">May 17, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<hr />
<h1>tl;dr</h1>
<ul>
<li>I continue to be alive and bad at both blog and implementing experimental data structures</li>
<li>I fixed my program not working by fixing the thing that told me it wasn&#8217;t working</li>
<li>If your evaluator has holes in, you&#8217;ll spend weeks chasing problems that don&#8217;t exist</li>
<li>Never assume how someone else&#8217;s software will work, especially if you are assuming it will work like a different piece of software that you are already making assumptions about</li>
<li>Always be testing (especially testing the testing)</li>
<li>This thing actually fucking works</li>
<li>An unpleasant side effect of doing a PhD is the rate of observed existential crises increases</li>
<li>Life continues to be a series of off-by-one errors, punctuated with occasional trips to the seaside</li>
</ul>
<div class="footnotes">
<hr />
<ol>
<li id="fn-699-1">
Let&#8217;s not even talk about indels for now.&#160;<a href="#fnref-699-1" rev="footnote">&#8617;</a> <a href="#fnref2:699-1" rev="footnote">&#8617;</a>
</li>
</ol>
</div>
</div>
                
                        <div class="tag-well"><span><span class="fa fa-tags"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/tag/graph/" rel="tag">graph</a>,&nbsp; <a href="https://samnicholls.net/tag/metagenome/" rel="tag">metagenome</a>,&nbsp; <a href="https://samnicholls.net/tag/metahaplome/" rel="tag">metahaplome</a>,&nbsp; <a href="https://samnicholls.net/tag/sitrep/" rel="tag">sitrep</a>,&nbsp; <a href="https://samnicholls.net/tag/status-report/" rel="tag">status report</a></span></div>
                
                <nav class="further-reading">
	<div class="previous">
		<span>Previous Post</span>
		<a href="https://samnicholls.net/2016/06/06/fail2ban-apache-dos/">Using fail2ban to mitigate simple DOS attacks against apache (or why I am a terrible sysop)</a>
	</div>
	<div class="next">
		<span>Next Post</span>
		<a href="https://samnicholls.net/2016/06/15/how-to-sphinx-readthedocs/">An idiot&#8217;s guide to Python documentation with Sphinx and ReadTheDocs</a>
	</div>
</nav>            </article>
                                <div class="card">

<div id="disqus_thread"></div>
                </div>
    
</section>

        <!-- Sidebar -->
        <aside class="col-md-4 sidebar-container">
		            <div id="widget-area" class="widget-area" role="complementary">
                <div id="custom_html-3" class="widget_text card widget widget_custom_html"><div class="textwidget custom-html-widget"><img src="https://pbs.twimg.com/profile_images/1044255545610436609/wt33ZbtU_400x400.jpg" height=120 style="margin-top: 20px; margin-right:10px; float: left;" alt="Sam Nicholls" />
<div style="padding-top:15px; line-height: normal"><b style="font-size: 1.2em">Sam Nicholls</b><br><span style="font-size: 0.9em;">Haplotype Wrangler at University of Birmingham.<br>I tell computers to do things, then I write about how it all went wrong.</span></div></div></div><div id="search-2" class="card widget widget_search"><form style="margin-top:15px" action="https://samnicholls.net/" id="searchform" method="get" role="form">
    <div class="form-group">
        <input type="text" class="form-control" name="s" id="s" placeholder="e.g. Search...">
    </div>
</form>
</div><div id="twitter_timeline-5" class="card widget widget_twitter_timeline"><a class="twitter-timeline" data-height="400" data-theme="light" data-link-color="#f96e5b" data-border-color="#e8e8e8" data-lang="EN" data-partner="jetpack" data-chrome="noheader nofooter noborders" href="https://twitter.com/samstudio8">My Tweets</a></div><div id="categories-5" class="card widget widget_categories"><h2>Categories</h2>
			<ul>
					<li class="cat-item cat-item-2"><a href="https://samnicholls.net/category/bioinf/">Bioinformatics</a>
<ul class='children'>
	<li class="cat-item cat-item-8"><a href="https://samnicholls.net/category/bioinf/au-phd/">AU-PhD</a>
</li>
	<li class="cat-item cat-item-3"><a href="https://samnicholls.net/category/bioinf/sanger-qc/" title="Sanger QC Project">Sanger-QC</a>
</li>
	<li class="cat-item cat-item-61"><a href="https://samnicholls.net/category/bioinf/tools/">Tools</a>
	<ul class='children'>
	<li class="cat-item cat-item-185"><a href="https://samnicholls.net/category/bioinf/tools/chitin/">chitin</a>
</li>
	</ul>
</li>
</ul>
</li>
	<li class="cat-item cat-item-80"><a href="https://samnicholls.net/category/devops/">Devops</a>
</li>
	<li class="cat-item cat-item-114"><a href="https://samnicholls.net/category/event/">Event</a>
</li>
	<li class="cat-item cat-item-21"><a href="https://samnicholls.net/category/gadgets/">Gadgets</a>
</li>
	<li class="cat-item cat-item-176"><a href="https://samnicholls.net/category/hardware/">Hardware</a>
<ul class='children'>
	<li class="cat-item cat-item-223"><a href="https://samnicholls.net/category/hardware/beehive-monitor/">Beehive Monitor</a>
</li>
	<li class="cat-item cat-item-178"><a href="https://samnicholls.net/category/hardware/esp8266/">ESP8266</a>
</li>
	<li class="cat-item cat-item-236"><a href="https://samnicholls.net/category/hardware/lego-sequencer/">Lego Sequencer</a>
</li>
	<li class="cat-item cat-item-177"><a href="https://samnicholls.net/category/hardware/lightstick/">Lightstick</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-151"><a href="https://samnicholls.net/category/lab/">Lab</a>
<ul class='children'>
	<li class="cat-item cat-item-150"><a href="https://samnicholls.net/category/lab/protocol-guides/">Protocol Guides</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-5"><a href="https://samnicholls.net/category/meta/">Meta</a>
</li>
	<li class="cat-item cat-item-62"><a href="https://samnicholls.net/category/mysteries/">Mysteries</a>
</li>
	<li class="cat-item cat-item-22"><a href="https://samnicholls.net/category/personal/">Personal</a>
</li>
	<li class="cat-item cat-item-4"><a href="https://samnicholls.net/category/photo/">Photography</a>
</li>
	<li class="cat-item cat-item-143"><a href="https://samnicholls.net/category/programming/">Programming</a>
<ul class='children'>
	<li class="cat-item cat-item-145"><a href="https://samnicholls.net/category/programming/documentation/">Documentation</a>
</li>
	<li class="cat-item cat-item-144"><a href="https://samnicholls.net/category/programming/python/">Python</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-48"><a href="https://samnicholls.net/category/status-report/">Status Report</a>
</li>
	<li class="cat-item cat-item-60"><a href="https://samnicholls.net/category/sysadm/">System Administration</a>
</li>
	<li class="cat-item cat-item-1"><a href="https://samnicholls.net/category/uncategorised/">Uncategorised</a>
</li>
			</ul>

			</div>            </div>
		        </aside>
    </div>

</div> <!-- /container -->

<footer id="main-footer" class="bg-primary">
	<div id="footer-widgets" class="container">
		    </div>
</footer>



	<div style="display:none">
	</div>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/comment-reply.min.js?ver=5.7.5' id='comment-reply-js'></script>
<script type='text/javascript' id='disqus_count-js-extra'>
/* <![CDATA[ */
var countVars = {"disqusShortname":"samposium-blog"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_count.js?ver=3.0.17' id='disqus_count-js'></script>
<script type='text/javascript' id='disqus_embed-js-extra'>
/* <![CDATA[ */
var embedVars = {"disqusConfig":{"integration":"wordpress 3.0.17"},"disqusIdentifier":"699 https:\/\/samnicholls.net\/?p=699","disqusShortname":"samposium-blog","disqusTitle":"Status Report: May 2016 (Metahaplomes: The graph that isn\u2019t)","disqusUrl":"https:\/\/samnicholls.net\/2016\/06\/12\/status-may16\/","postId":"699"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_embed.js?ver=3.0.17' id='disqus_embed-js'></script>
<script type='text/javascript' src='https://s0.wp.com/wp-content/js/devicepx-jetpack.js?ver=202335' id='devicepx-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/jetpack/_inc/twitter-timeline.js?ver=4.0.0' id='jetpack-twitter-timeline-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/wp-embed.min.js?ver=5.7.5' id='wp-embed-js'></script>
<script type='text/javascript' src='https://stats.wp.com/e-202335.js' async defer></script>
<script type='text/javascript'>
	_stq = window._stq || [];
	_stq.push([ 'view', {v:'ext',j:'1:4.9',blog:'101350222',post:'699',tz:'1',srv:'samnicholls.net'} ]);
	_stq.push([ 'clickTrackerInit', '101350222', '699' ]);
</script>

<script type="text/javascript">
(function() {
  if (navigator.userAgent.match(/IEMobile\/10\.0/)) {
    var msViewportStyle = document.createElement("style");
    msViewportStyle.appendChild(
      document.createTextNode("@-ms-viewport{width:auto!important}")
    );
    document.getElementsByTagName("head")[0].appendChild(msViewportStyle);
  }
})();

var collapsed = false;
jQuery( document ).ready(function($) {
    $('.navbar-toggle').click(function(){
        console.log(collapsed);
        if(collapsed) {
            $('.navbar-header > button').html('<i class="fa fa-bars"></i>');
        } else {
            $('.navbar-header > button').html('<i class="fa fa-times"></i>');
        }
        $('.navbar-collapse').toggle(200);
        collapsed = !collapsed;
    });
});
</script>
	
</body>

</html>
