<!DOCTYPE html>
<html lang="en-GB">

<head>
	<meta charset="UTF-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
	<link rel="pingback" href="https://samnicholls.net/xmlrpc.php">

	
	<title>bowtie2: Relaxed Parameters for Generous Alignments to Metagenomes &#8211; Samposium</title>
<meta name='robots' content='max-image-preview:large' />
<link rel='dns-prefetch' href='//s0.wp.com' />
<link rel='dns-prefetch' href='//fonts.googleapis.com' />
<link rel='dns-prefetch' href='//s.w.org' />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Feed" href="https://samnicholls.net/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; Comments Feed" href="https://samnicholls.net/comments/feed/" />
<link rel="alternate" type="application/rss+xml" title="Samposium &raquo; bowtie2: Relaxed Parameters for Generous Alignments to Metagenomes Comments Feed" href="https://samnicholls.net/2016/12/24/bowtie2-metagenomes/feed/" />
		<script type="text/javascript">
			window._wpemojiSettings = {"baseUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/72x72\/","ext":".png","svgUrl":"https:\/\/s.w.org\/images\/core\/emoji\/13.0.1\/svg\/","svgExt":".svg","source":{"concatemoji":"https:\/\/samnicholls.net\/wp-includes\/js\/wp-emoji-release.min.js?ver=5.7.5"}};
			!function(e,a,t){var n,r,o,i=a.createElement("canvas"),p=i.getContext&&i.getContext("2d");function s(e,t){var a=String.fromCharCode;p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,e),0,0);e=i.toDataURL();return p.clearRect(0,0,i.width,i.height),p.fillText(a.apply(this,t),0,0),e===i.toDataURL()}function c(e){var t=a.createElement("script");t.src=e,t.defer=t.type="text/javascript",a.getElementsByTagName("head")[0].appendChild(t)}for(o=Array("flag","emoji"),t.supports={everything:!0,everythingExceptFlag:!0},r=0;r<o.length;r++)t.supports[o[r]]=function(e){if(!p||!p.fillText)return!1;switch(p.textBaseline="top",p.font="600 32px Arial",e){case"flag":return s([127987,65039,8205,9895,65039],[127987,65039,8203,9895,65039])?!1:!s([55356,56826,55356,56819],[55356,56826,8203,55356,56819])&&!s([55356,57332,56128,56423,56128,56418,56128,56421,56128,56430,56128,56423,56128,56447],[55356,57332,8203,56128,56423,8203,56128,56418,8203,56128,56421,8203,56128,56430,8203,56128,56423,8203,56128,56447]);case"emoji":return!s([55357,56424,8205,55356,57212],[55357,56424,8203,55356,57212])}return!1}(o[r]),t.supports.everything=t.supports.everything&&t.supports[o[r]],"flag"!==o[r]&&(t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&t.supports[o[r]]);t.supports.everythingExceptFlag=t.supports.everythingExceptFlag&&!t.supports.flag,t.DOMReady=!1,t.readyCallback=function(){t.DOMReady=!0},t.supports.everything||(n=function(){t.readyCallback()},a.addEventListener?(a.addEventListener("DOMContentLoaded",n,!1),e.addEventListener("load",n,!1)):(e.attachEvent("onload",n),a.attachEvent("onreadystatechange",function(){"complete"===a.readyState&&t.readyCallback()})),(n=t.source||{}).concatemoji?c(n.concatemoji):n.wpemoji&&n.twemoji&&(c(n.twemoji),c(n.wpemoji)))}(window,document,window._wpemojiSettings);
		</script>
		<style type="text/css">
img.wp-smiley,
img.emoji {
	display: inline !important;
	border: none !important;
	box-shadow: none !important;
	height: 1em !important;
	width: 1em !important;
	margin: 0 .07em !important;
	vertical-align: -0.1em !important;
	background: none !important;
	padding: 0 !important;
}
</style>
	<link rel='stylesheet' id='crayon-css'  href='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/css/min/crayon.min.css?ver=_2.7.2_beta' type='text/css' media='all' />
<link rel='stylesheet' id='wp-block-library-css'  href='https://samnicholls.net/wp-includes/css/dist/block-library/style.min.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='child-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='lato-css'  href='//fonts.googleapis.com/css?family=Lato%3A400%2C700%2C400italic&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='slabo-css'  href='//fonts.googleapis.com/css?family=Slabo+27px&#038;ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='fa-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/font-awesome.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='rd-style-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe-child/style.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='css-custom-css'  href='https://samnicholls.net/wp-content/themes/responsive-deluxe/css/custom.css?ver=5.7.5' type='text/css' media='all' />
<link rel='stylesheet' id='jetpack_css-css'  href='https://samnicholls.net/wp-content/plugins/jetpack/css/jetpack.css?ver=4.9' type='text/css' media='all' />
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery.min.js?ver=3.5.1' id='jquery-core-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/jquery/jquery-migrate.min.js?ver=3.3.2' id='jquery-migrate-js'></script>
<script type='text/javascript' id='crayon_js-js-extra'>
/* <![CDATA[ */
var CrayonSyntaxSettings = {"version":"_2.7.2_beta","is_admin":"0","ajaxurl":"https:\/\/samnicholls.net\/wp-admin\/admin-ajax.php","prefix":"crayon-","setting":"crayon-setting","selected":"crayon-setting-selected","changed":"crayon-setting-changed","special":"crayon-setting-special","orig_value":"data-orig-value","debug":""};
var CrayonSyntaxStrings = {"copy":"Press %s to Copy, %s to Paste","minimize":"Click To Expand Code"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/crayon-syntax-highlighter/js/min/crayon.min.js?ver=_2.7.2_beta' id='crayon_js-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/fd-footnotes/fdfootnotes.js?ver=1.34' id='fdfootnote_script-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/bootstrap.js?ver=5.7.5' id='bootstrap-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/themes/responsive-deluxe/js/custom.js?ver=5.7.5' id='js-custom-js'></script>
<link rel="https://api.w.org/" href="https://samnicholls.net/wp-json/" /><link rel="alternate" type="application/json" href="https://samnicholls.net/wp-json/wp/v2/posts/1991" /><link rel="EditURI" type="application/rsd+xml" title="RSD" href="https://samnicholls.net/xmlrpc.php?rsd" />
<link rel="wlwmanifest" type="application/wlwmanifest+xml" href="https://samnicholls.net/wp-includes/wlwmanifest.xml" /> 
<meta name="generator" content="WordPress 5.7.5" />
<link rel="canonical" href="https://samnicholls.net/2016/12/24/bowtie2-metagenomes/" />
<link rel='shortlink' href='https://wp.me/p6RfP0-w7' />
<link rel="alternate" type="application/json+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F12%2F24%2Fbowtie2-metagenomes%2F" />
<link rel="alternate" type="text/xml+oembed" href="https://samnicholls.net/wp-json/oembed/1.0/embed?url=https%3A%2F%2Fsamnicholls.net%2F2016%2F12%2F24%2Fbowtie2-metagenomes%2F&#038;format=xml" />
<script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-62206786-1', 'auto');
  ga('send', 'pageview');

</script>


<link rel='dns-prefetch' href='//v0.wordpress.com'>
<style type='text/css'>img#wpstats{display:none}</style>
<!-- Jetpack Open Graph Tags -->
<meta property="og:type" content="article" />
<meta property="og:title" content="bowtie2: Relaxed Parameters for Generous Alignments to Metagenomes" />
<meta property="og:url" content="https://samnicholls.net/2016/12/24/bowtie2-metagenomes/" />
<meta property="og:description" content="Even `bowtie2`&#8217;s `&#8211;very-sensitive-local` option can be a little grumpy. Here&#8217;s how I make `bowtie2` a little more relaxed for alignment of more dissimilar sequences, particularly â€¦" />
<meta property="article:published_time" content="2016-12-24T00:34:46+00:00" />
<meta property="article:modified_time" content="2016-12-24T17:41:17+00:00" />
<meta property="og:site_name" content="Samposium" />
<meta property="og:image" content="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-23-00-41-15-1024x576.png" />
<meta property="og:locale" content="en_GB" />
<meta name="twitter:card" content="summary" />
</head>

<body class="post-template-default single single-post postid-1991 single-format-standard">


        
<header id="main-header">
    <nav class="navbar-default navbar-static-top navbar" role="navigation">
        <div class="container">
            <div class="navbar-header">
                <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#rd-collapse">
                    <span class="fa fa-bars"></span>
                    <span class="sr-only screen-reader-text">Toggle navigation</span>
                </button>
                <a class="navbar-brand" href=" https://samnicholls.net/">Samposium</a>            </div>
            
            <div id="rd-collapse" class="collapse navbar-collapse">
            
            <!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body class="post-template-default single single-post postid-1991 single-format-standard">

<div id="main-menu" class="navbar-nav nav"><ul>
<li class="page_item page-item-8"><a href="https://samnicholls.net/about/">About Me</a></li>
<li class="page_item page-item-11"><a href="https://samnicholls.net/talks/">Talks</a></li>
<li class="page_item page-item-15"><a href="https://samnicholls.net/things-i-have-learned/">Things I Have Learned</a></li>
</ul></div></body></html>
        </div>

        </div>
    </nav>
</header>

<div class="container">

    <div class="row"><!-- Content -->
<section class="col-md-8 content-container">
                          <article class="card post hentry post-1991 type-post status-publish format-standard category-tools tag-alignment tag-bowtie2 tag-config tag-loose tag-metagenome tag-metagenomics tag-options tag-parameters tag-reads tag-relax" id="post-1991">
                <header class="entry-header">
                        <h1 class="entry-title">bowtie2: Relaxed Parameters for Generous Alignments to Metagenomes</h1>                </header><!-- .entry-header -->

                <div class="meta-well">
                    <span style="margin-right:15px;" class=""><span class="fa fa-user"></span>&nbsp;&nbsp;&nbsp;<span class="author vcard"><a class="url fn n" href="https://samnicholls.net/about/">Sam Nicholls</a></span></span>

                    <span style="margin-right:15px;" class="entry-date published updated post-date"><span class="fa fa-clock-o"></span>&nbsp;&nbsp; <time class="entry-date published" datetime="2016-12-24T00:34:46+00:00">24th December 2016</time><time class="updated" datetime="2016-12-24T17:41:17+00:00">24th December 2016</time></span>                    <span style="margin-right:15px;" class=""><span class="fa fa-comments-o"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/2016/12/24/bowtie2-metagenomes/#respond"><span class="dsq-postid" data-dsqidentifier="1991 https://samnicholls.net/?p=1991">No Comments yet</span></a></span>
                    <span class=""><span class="fa fa-folder-open"></span>&nbsp;&nbsp;&nbsp;<a href="https://samnicholls.net/category/bioinf/tools/" rel="category tag">Tools</a></span>
                </div>

                <div class="entry-content"><p>In a change to my usual essay length posts, I wanted to share a quick <code>bowtie2</code> tip for relaxing the parameters of alignment. It&#8217;s no big secret that <code>bowtie2</code> has these options, and there&#8217;s some <a href="http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml">pretty good guidance in the manual</a>, too. However, we&#8217;ve had significant trouble in our lab finding a suitable set of permissive alignment parameters.</p>
<p><a href="https://samnicholls.net/category/bioinf/au-phd/">In the course of my PhD work on haplotyping regions of metagenomes</a>, I have found that even using <code>bowtie2</code>&#8216;s somewhat permissive <code>--very-sensitive-local</code>, that sequences with an identity to the reference of less than 90% are significantly less likely to align back to that reference. This is problematic in my line of work, where I wish to recover all of the individual variants of a gene, as the basis of my approach relies on a set of short reads (50-250bp) aligned to a position on a metagenomic assembly (that I term the <em>pseudo-reference</em>). It&#8217;s important to note that I am not interested in the assembly of individual genomes from metagenomic reads, but <strong>the genes themselves</strong>.</p>
<p>Recently, the opportunity arose to provide some evidence to this. I have some datasets which constitute &#8220;synthetic metahaplomes&#8221; that consist of a handful of arbitrary known genes that all perform the same function, each from a different organism. These genes can be broken up into synthetic reads and aligned to some common reference (another gene in the same family).</p>
<p>This alignment can be used a means to test <a href="https://samnicholls.net/2016/10/24/status-sep16/">my metagenomic haplotyper; <code>Gretel</code> (and her novel brother data structure, <code>Hansel</code>)</a>, by attempting to recover the original input sequences, from these synthetic reads. I&#8217;ve already <a href="http://biorxiv.org/content/early/2016/08/02/067215">reported in my pre-print</a> that our method is at the mercy of the preceding alignment, and used this as the hypothesis for a poor recovery in one of our data sets.</p>
<p>Indeed <a href="https://samnicholls.net/2016/12/24/status-nov16-p2/">as part of my latest experiments</a>, I have generated some coverage heat maps, showing the average coverage of each haplotype (Y-axis) at each position of the pseudo-reference (X-axis) and I&#8217;ve found that for sequences beyond the vicinity of 90% sequence identity, <code>--very-sensitive-local</code> becomes unsuitable.</p>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-22-18-40-53.png"><img src="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-22-18-40-53-1024x576.png" alt="" width="100%" class="alignnone size-large wp-image-2035" srcset="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-22-18-40-53-1024x576.png 1024w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-22-18-40-53-300x169.png 300w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-22-18-40-53-768x432.png 768w" sizes="(max-width: 1024px) 100vw, 1024px" /></a></p>
<p>The <code>BLAST</code> record below represents the alignment that corresponds to the gene whose reads go on to align at the average coverage depicted at the top bar of the above heatmap. Despite its 79% identity, it looks good(TM) to me, and I <strong>need</strong> sequence of this level of diversity to align to my pseudo-reference so it can be included in <code>Gretel</code>&#8216;s analysis. I need generous alignment parameters to permit even quite diverse reads (but hopefully not too diverse such that it is no longer a gene of the same family) to map back to my reference. Otherwise <code>Gretel</code> will simply miss these haplotypes.</p>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-24-00-14-39.png"><img src="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-24-00-14-39-1024x941.png" alt="" width="90%" class="alignnone size-large wp-image-2050" srcset="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-24-00-14-39-1024x941.png 1024w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-24-00-14-39-300x276.png 300w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-24-00-14-39-768x706.png 768w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-24-00-14-39.png 1028w" sizes="(max-width: 1024px) 100vw, 1024px" /></a></p>
<p>So despite having already spent many days of my PhD repeatedly failing to increase my overall alignment rates for my metagenomes, I felt this time it would be different. I had a method (my heatmap) to see how my alignment parameters affected the alignment rates of reads on a <em>per-haplotype</em> basis. It&#8217;s also taken until now for me to quantify just what sort of sequences we are missing out on, courtesy of dropped reads.</p>
<p>I was determined to get this right.</p>
<blockquote class="twitter-tweet" data-lang="en">
<p lang="en" dir="ltr">Playing fucking mastermind with bowtie2 parameters <a href="https://t.co/yDHduJPBNM">pic.twitter.com/yDHduJPBNM</a></p>
<p>&mdash; Sam Nicholls (@samstudio8) <a href="https://twitter.com/samstudio8/status/812099556200640512">December 23, 2016</a></p></blockquote>
<p><script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script></p>
<p>For a change, I&#8217;ll save you the anticipation and tell you what I settled on after about 36 hours of getting cross.</p>
<ul>
<li><strong><code>--local -D 20 -R 3</code></strong><br />
Ensure we&#8217;re not performing end-to-end alignment (allow for soft clipping and the like), and borrow the most sensitive default &#8220;effort&#8221; parameters.</li>
<li><strong><code>-L 3</code></strong><br />
The seed substring length. Decreasing this from the default (<code>20 - 25</code>) to just <code>3</code> allows for a much more aggressive alignment, but adds computational cost. I actually had reasonably good results with <code>-L 11</code>, which might suit you if you have a much larger data set but still need to relax the aligner.</li>
<li><strong><code>-N 1</code></strong><br />
Permit a mismatch in the seed, because why not?</li>
<li><strong><code>--gbar 1</code></strong><br />
Has a small, but noticeable effect. Appears to thin the width of some of the coverage gap in the heatmap at the most stubborn sites.</li>
<li><strong><code>--mp 4</code></strong><br />
Reduces the maximum penalty that can be applied to a strongly supported (high quality) mismatch by a third (from the default value of <code>6</code>). The aggregate sum of these penalties are responsible for the dropping of reads. Along with the substring length, this had a significant influence on increasing my alignment rates. If your coverage stains are stubborn, you could decrease this again.</li>
</ul>
<h2>Tada.</h2>
<p><a href="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-23-00-41-15.png"><img src="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-23-00-41-15-1024x576.png" alt="" width="100%" class="alignnone size-large wp-image-2058" srcset="https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-23-00-41-15-1024x576.png 1024w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-23-00-41-15-300x169.png 300w, https://samnicholls.net/wp-content/uploads/2016/12/Screenshot-from-2016-12-23-00-41-15-768x432.png 768w" sizes="(max-width: 1024px) 100vw, 1024px" /></a></p>
<hr />
<h1>tl;dr</h1>
<ul>
<li><code>bowtie2 --local -D 20 -R 3 -L 3 -N 1 -p 8 --gbar 1 --mp 3</code></li>
</ul>
</div>
                
                        <div class="tag-well"><span><span class="fa fa-tags"></span>&nbsp;&nbsp; <a href="https://samnicholls.net/tag/alignment/" rel="tag">alignment</a>,&nbsp; <a href="https://samnicholls.net/tag/bowtie2/" rel="tag">bowtie2</a>,&nbsp; <a href="https://samnicholls.net/tag/config/" rel="tag">config</a>,&nbsp; <a href="https://samnicholls.net/tag/loose/" rel="tag">loose</a>,&nbsp; <a href="https://samnicholls.net/tag/metagenome/" rel="tag">metagenome</a>,&nbsp; <a href="https://samnicholls.net/tag/metagenomics/" rel="tag">metagenomics</a>,&nbsp; <a href="https://samnicholls.net/tag/options/" rel="tag">options</a>,&nbsp; <a href="https://samnicholls.net/tag/parameters/" rel="tag">parameters</a>,&nbsp; <a href="https://samnicholls.net/tag/reads/" rel="tag">reads</a>,&nbsp; <a href="https://samnicholls.net/tag/relax/" rel="tag">relax</a></span></div>
                
                <nav class="further-reading">
	<div class="previous">
		<span>Previous Post</span>
		<a href="https://samnicholls.net/2016/12/19/status-nov16-p1/">Status Report: November 2016 (Part I): Triviomes, Treeviomes &#038; Fuck Everything</a>
	</div>
	<div class="next">
		<span>Next Post</span>
		<a href="https://samnicholls.net/2016/12/24/status-nov16-p2/">Status Report: November 2016 (Part II): Revisiting the Synthetic Metahaplomes</a>
	</div>
</nav>            </article>
                                <div class="card">

<div id="disqus_thread"></div>
                </div>
    
</section>

        <!-- Sidebar -->
        <aside class="col-md-4 sidebar-container">
		            <div id="widget-area" class="widget-area" role="complementary">
                <div id="custom_html-3" class="widget_text card widget widget_custom_html"><div class="textwidget custom-html-widget"><img src="https://pbs.twimg.com/profile_images/1044255545610436609/wt33ZbtU_400x400.jpg" height=120 style="margin-top: 20px; margin-right:10px; float: left;" alt="Sam Nicholls" />
<div style="padding-top:15px; line-height: normal"><b style="font-size: 1.2em">Sam Nicholls</b><br><span style="font-size: 0.9em;">Haplotype Wrangler at University of Birmingham.<br>I tell computers to do things, then I write about how it all went wrong.</span></div></div></div><div id="search-2" class="card widget widget_search"><form style="margin-top:15px" action="https://samnicholls.net/" id="searchform" method="get" role="form">
    <div class="form-group">
        <input type="text" class="form-control" name="s" id="s" placeholder="e.g. Search...">
    </div>
</form>
</div><div id="twitter_timeline-5" class="card widget widget_twitter_timeline"><a class="twitter-timeline" data-height="400" data-theme="light" data-link-color="#f96e5b" data-border-color="#e8e8e8" data-lang="EN" data-partner="jetpack" data-chrome="noheader nofooter noborders" href="https://twitter.com/samstudio8">My Tweets</a></div><div id="categories-5" class="card widget widget_categories"><h2>Categories</h2>
			<ul>
					<li class="cat-item cat-item-2"><a href="https://samnicholls.net/category/bioinf/">Bioinformatics</a>
<ul class='children'>
	<li class="cat-item cat-item-8"><a href="https://samnicholls.net/category/bioinf/au-phd/">AU-PhD</a>
</li>
	<li class="cat-item cat-item-3"><a href="https://samnicholls.net/category/bioinf/sanger-qc/" title="Sanger QC Project">Sanger-QC</a>
</li>
	<li class="cat-item cat-item-61"><a href="https://samnicholls.net/category/bioinf/tools/">Tools</a>
	<ul class='children'>
	<li class="cat-item cat-item-185"><a href="https://samnicholls.net/category/bioinf/tools/chitin/">chitin</a>
</li>
	</ul>
</li>
</ul>
</li>
	<li class="cat-item cat-item-80"><a href="https://samnicholls.net/category/devops/">Devops</a>
</li>
	<li class="cat-item cat-item-114"><a href="https://samnicholls.net/category/event/">Event</a>
</li>
	<li class="cat-item cat-item-21"><a href="https://samnicholls.net/category/gadgets/">Gadgets</a>
</li>
	<li class="cat-item cat-item-176"><a href="https://samnicholls.net/category/hardware/">Hardware</a>
<ul class='children'>
	<li class="cat-item cat-item-223"><a href="https://samnicholls.net/category/hardware/beehive-monitor/">Beehive Monitor</a>
</li>
	<li class="cat-item cat-item-178"><a href="https://samnicholls.net/category/hardware/esp8266/">ESP8266</a>
</li>
	<li class="cat-item cat-item-236"><a href="https://samnicholls.net/category/hardware/lego-sequencer/">Lego Sequencer</a>
</li>
	<li class="cat-item cat-item-177"><a href="https://samnicholls.net/category/hardware/lightstick/">Lightstick</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-151"><a href="https://samnicholls.net/category/lab/">Lab</a>
<ul class='children'>
	<li class="cat-item cat-item-150"><a href="https://samnicholls.net/category/lab/protocol-guides/">Protocol Guides</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-5"><a href="https://samnicholls.net/category/meta/">Meta</a>
</li>
	<li class="cat-item cat-item-62"><a href="https://samnicholls.net/category/mysteries/">Mysteries</a>
</li>
	<li class="cat-item cat-item-22"><a href="https://samnicholls.net/category/personal/">Personal</a>
</li>
	<li class="cat-item cat-item-4"><a href="https://samnicholls.net/category/photo/">Photography</a>
</li>
	<li class="cat-item cat-item-143"><a href="https://samnicholls.net/category/programming/">Programming</a>
<ul class='children'>
	<li class="cat-item cat-item-145"><a href="https://samnicholls.net/category/programming/documentation/">Documentation</a>
</li>
	<li class="cat-item cat-item-144"><a href="https://samnicholls.net/category/programming/python/">Python</a>
</li>
</ul>
</li>
	<li class="cat-item cat-item-48"><a href="https://samnicholls.net/category/status-report/">Status Report</a>
</li>
	<li class="cat-item cat-item-60"><a href="https://samnicholls.net/category/sysadm/">System Administration</a>
</li>
	<li class="cat-item cat-item-1"><a href="https://samnicholls.net/category/uncategorised/">Uncategorised</a>
</li>
			</ul>

			</div>            </div>
		        </aside>
    </div>

</div> <!-- /container -->

<footer id="main-footer" class="bg-primary">
	<div id="footer-widgets" class="container">
		    </div>
</footer>



	<div style="display:none">
	</div>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/comment-reply.min.js?ver=5.7.5' id='comment-reply-js'></script>
<script type='text/javascript' id='disqus_count-js-extra'>
/* <![CDATA[ */
var countVars = {"disqusShortname":"samposium-blog"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_count.js?ver=3.0.17' id='disqus_count-js'></script>
<script type='text/javascript' id='disqus_embed-js-extra'>
/* <![CDATA[ */
var embedVars = {"disqusConfig":{"integration":"wordpress 3.0.17"},"disqusIdentifier":"1991 https:\/\/samnicholls.net\/?p=1991","disqusShortname":"samposium-blog","disqusTitle":"bowtie2: Relaxed Parameters for Generous Alignments to Metagenomes","disqusUrl":"https:\/\/samnicholls.net\/2016\/12\/24\/bowtie2-metagenomes\/","postId":"1991"};
/* ]]> */
</script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/disqus-comment-system/public/js/comment_embed.js?ver=3.0.17' id='disqus_embed-js'></script>
<script type='text/javascript' src='https://s0.wp.com/wp-content/js/devicepx-jetpack.js?ver=202335' id='devicepx-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-content/plugins/jetpack/_inc/twitter-timeline.js?ver=4.0.0' id='jetpack-twitter-timeline-js'></script>
<script type='text/javascript' src='https://samnicholls.net/wp-includes/js/wp-embed.min.js?ver=5.7.5' id='wp-embed-js'></script>
<script type='text/javascript' src='https://stats.wp.com/e-202335.js' async defer></script>
<script type='text/javascript'>
	_stq = window._stq || [];
	_stq.push([ 'view', {v:'ext',j:'1:4.9',blog:'101350222',post:'1991',tz:'1',srv:'samnicholls.net'} ]);
	_stq.push([ 'clickTrackerInit', '101350222', '1991' ]);
</script>

<script type="text/javascript">
(function() {
  if (navigator.userAgent.match(/IEMobile\/10\.0/)) {
    var msViewportStyle = document.createElement("style");
    msViewportStyle.appendChild(
      document.createTextNode("@-ms-viewport{width:auto!important}")
    );
    document.getElementsByTagName("head")[0].appendChild(msViewportStyle);
  }
})();

var collapsed = false;
jQuery( document ).ready(function($) {
    $('.navbar-toggle').click(function(){
        console.log(collapsed);
        if(collapsed) {
            $('.navbar-header > button').html('<i class="fa fa-bars"></i>');
        } else {
            $('.navbar-header > button').html('<i class="fa fa-times"></i>');
        }
        $('.navbar-collapse').toggle(200);
        collapsed = !collapsed;
    });
});
</script>
	
</body>

</html>
